<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Read alignments &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Features" href="features.html" />
    <link rel="prev" title="Genomic intervals and genomic arrays" href="genomic.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="features.html" title="Features"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="genomic.html" title="Genomic intervals and genomic arrays"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="read-alignments">
<span id="alignments"></span><h1>Read alignments<a class="headerlink" href="#read-alignments" title="Permalink to this headline">¶</a></h1>
<div class="section" id="concepts">
<h2>Concepts<a class="headerlink" href="#concepts" title="Permalink to this headline">¶</a></h2>
<p>There are a large number of different tools to align short reads to a reference. Most of
them use their own output format, even though the <a class="reference external" href="http://samtools.sourceforge.net/SAM1.pdf">SAM format</a> seems to become the common
standard now that many of the newer tools use it.</p>
<p>HTSeq aims to offer a uniform way to analyse alignments from different tools. To this end,
for all supported alignment formats a parse class is offered that reads an alignment file
and generates an iterator over the individual alignment records. These are represented as
objects of a sub-class of <a class="reference internal" href="#HTSeq.Alignment" title="HTSeq.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and hence all offer a common interface.</p>
<p>So, you can easily write code that should work for all aligner formats. As a simple example,
consider this function that counts the number of reads falling on each chromosome:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">collections</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">count_in_chroms</span><span class="p">(</span> <span class="n">alignments</span> <span class="p">):</span>
<span class="gp">... </span>    <span class="n">counts</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span> <span class="k">lambda</span><span class="p">:</span> <span class="mi">0</span> <span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="n">almnt</span><span class="o">.</span><span class="n">aligned</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">counts</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">chrom</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">counts</span>
</pre></div>
</div>
<p>If you have a SAM file (e.g., from BWA or BowTie), you can call it with:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">count_in_chroms</span><span class="p">(</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">SAM_Reader</span><span class="p">(</span> <span class="s2">&quot;yeast_RNASeq_excerpt.sam&quot;</span> <span class="p">)</span> <span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">())</span> 
<span class="go">[(&#39;2-micron&#39;, 46), (&#39;I&#39;, 362), (&#39;II&#39;, 1724), (&#39;III&#39;, 365), (&#39;IV&#39;, 3015),</span>
<span class="go"> (&#39;IX&#39;, 648), (&#39;V&#39;, 999), (&#39;VI&#39;, 332), (&#39;VII&#39;, 2316), (&#39;VIII&#39;, 932),</span>
<span class="go"> (&#39;X&#39;, 1129), (&#39;XI&#39;, 1170), (&#39;XII&#39;, 4215), (&#39;XIII&#39;, 1471), (&#39;XIV&#39;, 1297),</span>
<span class="go"> (&#39;XV&#39;, 2133), (&#39;XVI&#39;, 1509)]</span>
</pre></div>
</div>
<p>If, however, you have done your alignment with Eland from the SolexaPipeline, which
uses the &#8220;Solexa export&#8221; format, you can use the same function, only using <a class="reference internal" href="#HTSeq.SolexaExportReader" title="HTSeq.SolexaExportReader"><code class="xref py py-class docutils literal"><span class="pre">SolexaExportReader</span></code></a>
instead of <a class="reference internal" href="#HTSeq.SAM_Reader" title="HTSeq.SAM_Reader"><code class="xref py py-class docutils literal"><span class="pre">SAM_Reader</span></code></a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">count_in_chroms</span><span class="p">(</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">SolexaExportReader</span><span class="p">(</span> <span class="s2">&quot;mydata_export.txt&quot;</span> <span class="p">)</span> <span class="p">)</span> 
</pre></div>
</div>
<p>Both class generate iterators of similar objects. On the other hand, some formats contain more information
and then the <code class="docutils literal"><span class="pre">Alignment</span></code> objects from these contain additional fields.</p>
</div>
<div class="section" id="parser-classes">
<h2>Parser classes<a class="headerlink" href="#parser-classes" title="Permalink to this headline">¶</a></h2>
<p>Depending on the format of your alignment file, choose from the following parsers:</p>
<dl class="class">
<dt id="HTSeq.BowtieReader">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">BowtieReader</code><span class="sig-paren">(</span><em>filename_or_sequence</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.BowtieReader" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.SAM_Reader">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">SAM_Reader</code><span class="sig-paren">(</span><em>filename_or_sequence</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SAM_Reader" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.BAM_Reader">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">BAM_Reader</code><span class="sig-paren">(</span><em>filename_or_sequence</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.BAM_Reader" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.SolexaExportReader">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">SolexaExportReader</code><span class="sig-paren">(</span><em>filename_or_sequence</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SolexaExportReader" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>All of these are derived from <a class="reference internal" href="misc.html#HTSeq.FileOrSequence" title="HTSeq.FileOrSequence"><code class="xref py py-class docutils literal"><span class="pre">FileOrSequence</span></code></a>. When asked for an iterator,
they yield <code class="docutils literal"><span class="pre">Alignment</span></code> objects of types <a class="reference internal" href="#HTSeq.BowtieAlignment" title="HTSeq.BowtieAlignment"><code class="xref py py-class docutils literal"><span class="pre">BowtieAlignment</span></code></a>, <a class="reference internal" href="#HTSeq.SAM_Alignment" title="HTSeq.SAM_Alignment"><code class="xref py py-class docutils literal"><span class="pre">SAM_Alignment</span></code></a>,
or <a class="reference internal" href="#HTSeq.SolexaExportAlignment" title="HTSeq.SolexaExportAlignment"><code class="xref py py-class docutils literal"><span class="pre">SolexaExportAlignment</span></code></a>. See below for their properties.</p>
<p>Adding support for a new format is very easy. Ask me if you need something and
I can probably add it right-away.
Alternatively, you can convert your format to the SAM format. The <a class="reference external" href="http://samtools.sourceforge.net/">SAMtools</a>
contain Perl skripts to convert nearly all common formats.</p>
<blockquote>
<div><dl class="method">
<dt>
<code class="descname">SAM_Reader.peek( num = 1 ):</code></dt>
<dd><p>Peek into a SAM file or connection, reporting the first <code class="docutils literal"><span class="pre">num</span></code> records.
If you then call an iterator on the <code class="docutils literal"><span class="pre">SAM_Reader</span></code>, the record will
be yielded again.</p>
</dd></dl>

</div></blockquote>
</div>
<div class="section" id="alignment-and-alignmentwithsequencereversal">
<h2><code class="docutils literal"><span class="pre">Alignment</span></code> and <code class="docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code><a class="headerlink" href="#alignment-and-alignmentwithsequencereversal" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="HTSeq.Alignment">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">Alignment</code><span class="sig-paren">(</span><em>read</em>, <em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the base class of all Alignment classes. Any class derived
from <code class="docutils literal"><span class="pre">Alignment</span></code> has at least the following attributes:</p>
<dl class="attribute">
<dt id="HTSeq.Alignment.read">
<code class="descname">read</code><a class="headerlink" href="#HTSeq.Alignment.read" title="Permalink to this definition">¶</a></dt>
<dd><p>The read. An object of type <code class="xref py py-class docutils literal"><span class="pre">SequenceWithQuality</span></code>. See there for the sub-attributes.</p>
<p>Note that some aligners store the reverse complement of the read if it was
aligned to the &#8216;-&#8216; strand. In this case, the parser revers-complements the read
again, so that you can be sure that the read is always presented as it was sequenced
(see also <a class="reference internal" href="#HTSeq.AlignmentWithSequenceReversal" title="HTSeq.AlignmentWithSequenceReversal"><code class="xref py py-class docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code></a>).</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.Alignment.aligned">
<code class="descname">aligned</code><a class="headerlink" href="#HTSeq.Alignment.aligned" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Some formats (e.g., those of Maq and Bowtie) contain only aligned
reads (and the aligner collects the
unaligned reads in a seperate FASTQ file if requested). For these formats, <code class="docutils literal"><span class="pre">aligned</span></code>
is always <code class="docutils literal"><span class="pre">True</span></code>. Other formats (e.g., SAM and Solexa Export) list all reads, including those which could
not be aligned. In that case, check <code class="docutils literal"><span class="pre">aligned</span></code> to see whether the read has an
alignment.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.Alignment.iv">
<code class="descname">iv</code><a class="headerlink" href="#HTSeq.Alignment.iv" title="Permalink to this definition">¶</a></dt>
<dd><p>An object of class <a class="reference internal" href="genomic.html#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> or <code class="docutils literal"><span class="pre">None</span></code>.</p>
<p>The genomic interval to which the read was aligned (or <code class="docutils literal"><span class="pre">None</span></code> if <code class="docutils literal"><span class="pre">aligned=False</span></code>).
See <a class="reference internal" href="genomic.html#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> for the sub-attributes. Note that different formats
have different conventions for genomic coordinates. The parser class takes care
of normalizing this, so that <code class="docutils literal"><span class="pre">iv</span></code> always adheres to the conventions outlined
in :class:GenomicInterval. Especially, all coordinates are counted from zero, not one.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.Alignment.paired_end">
<code class="descname">paired_end</code><a class="headerlink" href="#HTSeq.Alignment.paired_end" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. True if the read stems from a paired-end sequencing run. (Note: At the moment
paired-end data is only supported for the SAM format.)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="HTSeq.AlignmentWithSequenceReversal">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">AlignmentWithSequenceReversal</code><span class="sig-paren">(</span><em>read_as_aligned</em>, <em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.AlignmentWithSequenceReversal" title="Permalink to this definition">¶</a></dt>
<dd><p>Some aligners store the reverse complement of the read if it was
aligned to the &#8216;-&#8216; strand. For these aligners, the Alignment class is derived
from <code class="docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code>, which undoes the reverse-complement if necessary
to ensure that the <code class="docutils literal"><span class="pre">read</span></code> attribute always presents the read in the ordder in which
it was sequenced.</p>
<p>To get better performance, this is done via lazy evaluation, i.e., the
reverse complement is only calculated when the <code class="docutils literal"><span class="pre">read</span></code> attribute is accessed
for the first time. The original read as read from the file is stored as well. You
can access both with these attributes:</p>
<dl class="attribute">
<dt id="HTSeq.AlignmentWithSequenceReversal.read_as_aligned">
<code class="descname">read_as_aligned</code><a class="headerlink" href="#HTSeq.AlignmentWithSequenceReversal.read_as_aligned" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="sequences.html#HTSeq.SequenceWithQualities" title="HTSeq.SequenceWithQualities"><code class="xref py py-class docutils literal"><span class="pre">SequenceWithQualities</span></code></a> object. The read as it was found in the file.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.AlignmentWithSequenceReversal.read_as_sequenced">
<code class="descname">read_as_sequenced</code><a class="headerlink" href="#HTSeq.AlignmentWithSequenceReversal.read_as_sequenced" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="sequences.html#HTSeq.SequenceWithQualities" title="HTSeq.SequenceWithQualities"><code class="xref py py-class docutils literal"><span class="pre">SequenceWithQualities</span></code></a> object. The read as it was sequenced,
i.e., an alias for <a class="reference internal" href="#HTSeq.Alignment.read" title="HTSeq.Alignment.read"><code class="xref py py-attr docutils literal"><span class="pre">Alignment.read</span></code></a>.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="format-specific-alignment-classes">
<h2>Format-specific Alignment classes<a class="headerlink" href="#format-specific-alignment-classes" title="Permalink to this headline">¶</a></h2>
<p>Note: All format-specific Alignment classes take a string as argument for their constructor. This
is a line from the alignment file describing the alignment and is passed in by the corresponding
<code class="docutils literal"><span class="pre">Reader</span></code> object. As you do not create <code class="docutils literal"><span class="pre">Alignment</span></code> objects yourself but get them from the <code class="docutils literal"><span class="pre">Reader</span></code>
object you typically never call the constructor yourself.</p>
<dl class="class">
<dt id="HTSeq.BowtieAlignment">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">BowtieAlignment</code><span class="sig-paren">(</span><em>bowtie_line</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.BowtieAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">BowtieAlignment</span></code> objects contain all the attributes from <a class="reference internal" href="#HTSeq.Alignment" title="HTSeq.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and
<a class="reference internal" href="#HTSeq.AlignmentWithSequenceReversal" title="HTSeq.AlignmentWithSequenceReversal"><code class="xref py py-class docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code></a>, and, in addition, these:</p>
<dl class="attribute">
<dt id="HTSeq.BowtieAlignment.reserved">
<code class="descname">reserved</code><a class="headerlink" href="#HTSeq.BowtieAlignment.reserved" title="Permalink to this definition">¶</a></dt>
<dd><p>A string. The <code class="docutils literal"><span class="pre">reserved</span></code> field from the Bowtie output file. See the Bowtie manual for its meaning.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.BowtieAlignment.substitutions">
<code class="descname">substitutions</code><a class="headerlink" href="#HTSeq.BowtieAlignment.substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>A string. The substitutions string that describes mismatches in the format <code class="docutils literal"><span class="pre">22:A&gt;C,</span> <span class="pre">25:C&gt;T</span></code>
to indicate a change from A to C in position 22 and from C to T in position 25.
No further parsing for this is offered yet.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="HTSeq.SAM_Alignment">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">SAM_Alignment</code><span class="sig-paren">(</span><em>line</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SAM_Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">BowtieAlignment</span></code> objects contain all the attributes from <a class="reference internal" href="#HTSeq.Alignment" title="HTSeq.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and
<a class="reference internal" href="#HTSeq.AlignmentWithSequenceReversal" title="HTSeq.AlignmentWithSequenceReversal"><code class="xref py py-class docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code></a>, and, in addition, these:</p>
<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.aQual">
<code class="descname">aQual</code><a class="headerlink" href="#HTSeq.SAM_Alignment.aQual" title="Permalink to this definition">¶</a></dt>
<dd><p>An int. The alignment quality score in Phread style encoding.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.cigar">
<code class="descname">cigar</code><a class="headerlink" href="#HTSeq.SAM_Alignment.cigar" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of <a class="reference internal" href="#HTSeq.CigarOperation" title="HTSeq.CigarOperation"><code class="xref py py-class docutils literal"><span class="pre">CigarOperation</span></code></a> objects, as parsed from the extended CIGAR string. See
<a class="reference internal" href="#HTSeq.CigarOperation" title="HTSeq.CigarOperation"><code class="xref py py-class docutils literal"><span class="pre">CigarOperation</span></code></a> for details.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.nor_primary_alignment">
<code class="descname">nor_primary_alignment</code><a class="headerlink" href="#HTSeq.SAM_Alignment.nor_primary_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Whether the alignment is not primary. (See SAM format reference, flag 0x0100.)</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.failed_platform_qc">
<code class="descname">failed_platform_qc</code><a class="headerlink" href="#HTSeq.SAM_Alignment.failed_platform_qc" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Whether the read failed a platform quality check. (See SAM format reference, flag 0x0200.)</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.pcr_or_optical_duplicate">
<code class="descname">pcr_or_optical_duplicate</code><a class="headerlink" href="#HTSeq.SAM_Alignment.pcr_or_optical_duplicate" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Whether the read is a PCR or optical duplicate. (See SAM format reference, flag 0x0400.)</p>
</dd></dl>

<p>These methods access the optional fields:</p>
<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.optional_field">
<code class="descname">optional_field</code><span class="sig-paren">(</span><em>tag</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SAM_Alignment.optional_field" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the optional field <code class="docutils literal"><span class="pre">tag</span></code>. See SAM format reference for the defined tags (which
are two-letter strings).</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.optional_fields">
<code class="descname">optional_fields</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SAM_Alignment.optional_fields" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dict with all optional fields, using their tags as keys.</p>
</dd></dl>

<p>This method is useful to write out a SAM file:</p>
<dl class="method">
<dt id="HTSeq.SAM_Alignment.get_sam_line">
<code class="descname">get_sam_line</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SAM_Alignment.get_sam_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Constructs a SAM line to describe the alignment, which is returned as a string.</p>
</dd></dl>

<p><strong>Paired-end support</strong></p>
<blockquote>
<div><p>SAM_Alignment objects can represent paired-end data. If <a class="reference internal" href="#HTSeq.Alignment.paired_end" title="HTSeq.Alignment.paired_end"><code class="xref py py-attr docutils literal"><span class="pre">Alignment.paired_end</span></code></a> is True,
the following fields may be used:</p>
<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.mate_aligned">
<code class="descname">mate_aligned</code><a class="headerlink" href="#HTSeq.SAM_Alignment.mate_aligned" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Whether the mate was aligned</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.pe_which">
<code class="descname">pe_which</code><a class="headerlink" href="#HTSeq.SAM_Alignment.pe_which" title="Permalink to this definition">¶</a></dt>
<dd><p>A string. Takes one of the values &#8220;first&#8221;, &#8220;second&#8221;, &#8220;unknown&#8221; and &#8220;not_paired_end&#8221;, to indicate
whether the read stems from the first or second pass of the paired-end sequencing.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.proper_pair">
<code class="descname">proper_pair</code><a class="headerlink" href="#HTSeq.SAM_Alignment.proper_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Boolean. Whether the mates form a proper pair. (See SAM format reference, flag 0x0002.)</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.mate_start">
<code class="descname">mate_start</code><a class="headerlink" href="#HTSeq.SAM_Alignment.mate_start" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="genomic.html#HTSeq.GenomicPosition" title="HTSeq.GenomicPosition"><code class="xref py py-class docutils literal"><span class="pre">GenomicPosition</span></code></a> object. The start (i.e., left-most position) of the mate&#8217;s alignment.
Note that mate_start.strand is opposite to iv.strand for proper pairs.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SAM_Alignment.inferred_insert_size">
<code class="descname">inferred_insert_size</code><a class="headerlink" href="#HTSeq.SAM_Alignment.inferred_insert_size" title="Permalink to this definition">¶</a></dt>
<dd><p>An int. The inferred size of the insert between the reads.</p>
</dd></dl>

</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="HTSeq.pair_SAM_alignments">
<code class="descclassname">HTSeq.</code><code class="descname">pair_SAM_alignments</code><span class="sig-paren">(</span><em>alnmt_seq</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.pair_SAM_alignments" title="Permalink to this definition">¶</a></dt>
<dd><p>This function takes a generator of <a class="reference internal" href="#HTSeq.SAM_Alignment" title="HTSeq.SAM_Alignment"><code class="xref py py-class docutils literal"><span class="pre">SAM_Alignment</span></code></a> objects (e.g.,
a <a class="reference internal" href="#HTSeq.SAM_Reader" title="HTSeq.SAM_Reader"><code class="xref py py-class docutils literal"><span class="pre">SAM_Reader</span></code></a> object) and yields a sequence of pairs of alignments.
A typical use may be:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="ow">in</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">SAM_Reader</span><span class="p">(</span> <span class="s2">&quot;some_paired_end_data.sam&quot;</span> <span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pair, consisting of&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;   &quot;</span><span class="p">,</span> <span class="n">first</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;   &quot;</span><span class="p">,</span> <span class="n">second</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, <code class="docutils literal"><span class="pre">first</span></code> and <code class="docutils literal"><span class="pre">second</span></code> are <a class="reference internal" href="#HTSeq.SAM_Alignment" title="HTSeq.SAM_Alignment"><code class="xref py py-class docutils literal"><span class="pre">SAM_Alignment</span></code></a> objects, representing two reads
of the same cluster. For this to work, the SAM file has to be arranged such that
paired reads are always in adjacent lines. As the SAM format requires that the query names
(first column of the SAM file) is the same for mate pairs, this arrangement can easily be
achieved by sorting the SAM file lines lexicographically.</p>
<p>Special care is taken to properly pair up multiple alignment lines for the same read.</p>
<p>In the SAM format, alignments for paired-end reads must be reported in paired alignment
records. If the mate of an alignment record is missing, this fact is counted and, at the
end, a warning stating the number of such violating reads is issued. The singleton
alignments are yielded as pairs, with the alignment in the first or second
element of the pair (depending on the sequencing pass it originates from) and the other
element is set to <code class="docutils literal"><span class="pre">None</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="HTSeq.pair_SAM_alignments_with_buffer">
<code class="descclassname">HTSeq.</code><code class="descname">pair_SAM_alignments_with_buffer</code><span class="sig-paren">(</span><em>alignments</em>, <em>max_buffer_size=3000000</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.pair_SAM_alignments_with_buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>This function pairs up reads in a SAM file, in the same manner as <a class="reference internal" href="#HTSeq.pair_SAM_alignments" title="HTSeq.pair_SAM_alignments"><code class="xref py py-func docutils literal"><span class="pre">pair_SAM_alignments()</span></code></a>
but does not require that mated alignments appear in adjacent records, i.e., the SAM
file does not need to be sorted by read name beforehand. Rather, once the first alignment
of a pair is encountered, it is stored in a buffer until its mated alignment is encountered,
and then both are yielded together as pair. It is recommended that the data should be
sorted by position, because then, mated alignments will typicalle not be too distant from
each other in the file and hence only a limited number of alignments have to be held
concurrently in the buffer, thereby reducing memory needs. To avoid overflowing the
system&#8217;s memory, the function stops and raises an exception once the number of
alignment records held in the buffer exceeds <code class="docutils literal"><span class="pre">max_buffer_size</span></code>.</p>
</dd></dl>

<dl class="class">
<dt id="HTSeq.SolexaExportAlignment">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">SolexaExportAlignment</code><span class="sig-paren">(</span><em>line</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.SolexaExportAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">SolexaExportAlignment</span></code> objects contain all the attributes from <a class="reference internal" href="#HTSeq.Alignment" title="HTSeq.Alignment"><code class="xref py py-class docutils literal"><span class="pre">Alignment</span></code></a> and
<a class="reference internal" href="#HTSeq.AlignmentWithSequenceReversal" title="HTSeq.AlignmentWithSequenceReversal"><code class="xref py py-class docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code></a>, and, in addition, these:</p>
<dl class="attribute">
<dt id="HTSeq.SolexaExportAlignment.passed_filter">
<code class="descname">passed_filter</code><a class="headerlink" href="#HTSeq.SolexaExportAlignment.passed_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. Whether the read passed the chastity filter. If <code class="docutils literal"><span class="pre">passed_filter==False</span></code>, then
<code class="docutils literal"><span class="pre">aligned==False</span></code>.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.SolexaExportAlignment.nomatch_code">
<code class="descname">nomatch_code</code><a class="headerlink" href="#HTSeq.SolexaExportAlignment.nomatch_code" title="Permalink to this definition">¶</a></dt>
<dd><p>A string. For <code class="docutils literal"><span class="pre">aligned==False</span></code>, a code indicating why no match could be found. See the
description of the 11th column of the Solexa Export format in the SolexaPipeline
manual for the meaning of the codes. For <code class="docutils literal"><span class="pre">aligned==True</span></code>, <code class="docutils literal"><span class="pre">nomatch_code==None</span></code>.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="multiple-alignments">
<h2>Multiple alignments<a class="headerlink" href="#multiple-alignments" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="HTSeq.bundle_multiple_alignments">
<code class="descclassname">HTSeq.</code><code class="descname">bundle_multiple_alignments</code><span class="sig-paren">(</span><em>sequence_of_alignments</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.bundle_multiple_alignments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>Some alignment programs, e.g., Bowtie, can output multiple alignments,
i.e., the same read is reported consecutively with different alignments.
This function takes an iterator over alignments (as provided by one of the
alignment Reader classes) and bundles consecutive alignments regarding the
same read to a list of Alignment objects and returns an iterator over these.</p>
</div>
<div class="section" id="cigar-strings">
<h2>CIGAR strings<a class="headerlink" href="#cigar-strings" title="Permalink to this headline">¶</a></h2>
<p>When reading in SAM files, the CIGAR string is parsed and stored as a list of
<code class="docutils literal"><span class="pre">CigarOperation</span></code> objects. For example, assume, a 36 bp read has been aligned
to the &#8216;+&#8217; strand of chromosome &#8216;chr3&#8217;, extending to the right from position
1000, with the CIGAR string <code class="docutils literal"><span class="pre">&quot;20M6I10M&quot;</span></code>. The function :function:parse_cigar
spells out what this means:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">HTSeq</span><span class="o">.</span><span class="n">parse_cigar</span><span class="p">(</span> <span class="s2">&quot;20M6I10M&quot;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span>  
<span class="go">[&lt; CigarOperation: 20 base(s) matched on ref iv chr2:[1000,1020)/+, query iv [0,20) &gt;,</span>
<span class="go"> &lt; CigarOperation: 6 base(s) inserted on ref iv chr2:[1020,1020)/+, query iv [20,26) &gt;,</span>
<span class="go"> &lt; CigarOperation: 10 base(s) matched on ref iv chr2:[1020,1030)/+, query iv [26,36) &gt;]</span>
</pre></div>
</div>
<p>We can see that the map includes an insert. Hence, the affected coordinates run from 1000
to 1030 on the reference (i.e., the chromosome) but only from 0 to 36 on the query (i.e., the read).</p>
<p>We can convenient access to the parsed data by looking at the attributes of the three <code class="docutils literal"><span class="pre">CigarOperation</span></code>
objects in the list.</p>
<dl class="class">
<dt id="HTSeq.CigarOperation">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">CigarOperation</code><span class="sig-paren">(</span><em>...</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.CigarOperation" title="Permalink to this definition">¶</a></dt>
<dd><p>The available attributes are:</p>
<dl class="attribute">
<dt id="HTSeq.CigarOperation.type">
<code class="descname">type</code><a class="headerlink" href="#HTSeq.CigarOperation.type" title="Permalink to this definition">¶</a></dt>
<dd><p>The type of the operation. One of the letters M, I, D, N, S, H, or P. Use
the dict <strong>cigar_operation_names</strong> to transform this to names:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">HTSeq</span><span class="o">.</span><span class="n">cigar_operation_names</span><span class="o">.</span><span class="n">items</span><span class="p">())</span>    
<span class="go">[(&#39;D&#39;, &#39;deleted&#39;),</span>
<span class="go"> (&#39;H&#39;, &#39;hard-clipped&#39;),</span>
<span class="go"> (&#39;I&#39;, &#39;inserted&#39;),</span>
<span class="go"> (&#39;M&#39;, &#39;matched&#39;),</span>
<span class="go"> (&#39;N&#39;, &#39;skipped&#39;),</span>
<span class="go"> (&#39;P&#39;, &#39;padded&#39;),</span>
<span class="go"> (&#39;S&#39;, &#39;soft-clipped&#39;)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.CigarOperation.size">
<code class="descname">size</code><a class="headerlink" href="#HTSeq.CigarOperation.size" title="Permalink to this definition">¶</a></dt>
<dd><p>The number of affected bases, an int.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.CigarOperation.ref_iv">
<code class="descname">ref_iv</code><a class="headerlink" href="#HTSeq.CigarOperation.ref_iv" title="Permalink to this definition">¶</a></dt>
<dd><p>A <a class="reference internal" href="genomic.html#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> specifying the affected bases on the reference. In case
of an insertion, this is a zero-length interval.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.CigarOperation.query_from">
<code class="descname">query_from</code><a class="headerlink" href="#HTSeq.CigarOperation.query_from" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.CigarOperation.query_to">
<code class="descname">query_to</code><a class="headerlink" href="#HTSeq.CigarOperation.query_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Two ints, specifying the affected bases on the query (the read). In case of a
deletion, <code class="docutils literal"><span class="pre">query_from</span> <span class="pre">==</span> <span class="pre">query_to</span></code>.</p>
</dd></dl>

<dl class="method">
<dt id="HTSeq.CigarOperation.check">
<code class="descname">check</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.CigarOperation.check" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the <code class="docutils literal"><span class="pre">CigarOperation</span></code> object for consitency. Returns a boolean.</p>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Read alignments</a><ul>
<li><a class="reference internal" href="#concepts">Concepts</a></li>
<li><a class="reference internal" href="#parser-classes">Parser classes</a></li>
<li><a class="reference internal" href="#alignment-and-alignmentwithsequencereversal"><code class="docutils literal"><span class="pre">Alignment</span></code> and <code class="docutils literal"><span class="pre">AlignmentWithSequenceReversal</span></code></a></li>
<li><a class="reference internal" href="#format-specific-alignment-classes">Format-specific Alignment classes</a></li>
<li><a class="reference internal" href="#multiple-alignments">Multiple alignments</a></li>
<li><a class="reference internal" href="#cigar-strings">CIGAR strings</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="genomic.html"
                        title="previous chapter">Genomic intervals and genomic arrays</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="features.html"
                        title="next chapter">Features</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/alignments.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="features.html" title="Features"
             >next</a> |</li>
        <li class="right" >
          <a href="genomic.html" title="Genomic intervals and genomic arrays"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>