<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Features &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Other parsers" href="otherparsers.html" />
    <link rel="prev" title="Read alignments" href="alignments.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="otherparsers.html" title="Other parsers"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="alignments.html" title="Read alignments"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="features">
<span id="id1"></span><h1>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<p>The easiest way to work with annotation is to use <a class="reference internal" href="genomic.html#HTSeq.GenomicArray" title="HTSeq.GenomicArray"><code class="xref py py-class docutils literal"><span class="pre">GenomicArray</span></code></a> with <code class="docutils literal"><span class="pre">typecode=='O'</span></code>
or <a class="reference internal" href="genomic.html#HTSeq.GenomicArrayOfSets" title="HTSeq.GenomicArrayOfSets"><code class="xref py py-class docutils literal"><span class="pre">GenomicArrayOfSets</span></code></a>. If you have your annotation in a flat file, with each
line describing a feature and giving its coordinates, you can read in the file line for line,
parse it (see the standard Python module <code class="docutils literal"><span class="pre">csv</span></code>), use the information on chromosome, start,
end and strand to create a <a class="reference internal" href="genomic.html#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> object and then store the data from the line
in the genomic array at the place indicated by the genomic interval.</p>
<p>For example, if you have data in a tab-separated file as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span> <span class="s2">&quot;feature_list.txt&quot;</span> <span class="p">):</span>  
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
<span class="go">chr2  100    300     +       &quot;gene A&quot;</span>
<span class="go">chr2 200     400     -       &quot;gene B&quot;</span>
<span class="go">chr3 150     270     +       &quot;gene C&quot;</span>
</pre></div>
</div>
<p>Then, you could load this information as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">csv</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">genes</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">(</span> <span class="p">[</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">,</span> <span class="s2">&quot;chr3&quot;</span> <span class="p">],</span> <span class="n">typecode</span><span class="o">=</span><span class="s1">&#39;O&#39;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="ow">in</span> \
<span class="gp">... </span>       <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;feature_list.txt&quot;</span><span class="p">),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">):</span>
<span class="gp">... </span>    <span class="n">iv</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">chrom</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">),</span> <span class="n">strand</span> <span class="p">)</span>
<span class="gp">... </span>    <span class="n">genes</span><span class="p">[</span> <span class="n">iv</span> <span class="p">]</span> <span class="o">=</span> <span class="n">name</span>
</pre></div>
</div>
<p>Now, to see whether there is a feature at a given <a class="reference internal" href="genomic.html#HTSeq.GenomicPosition" title="HTSeq.GenomicPosition"><code class="xref py py-class docutils literal"><span class="pre">GenomicPosition</span></code></a>, you just query the
genomic array:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">])</span>
<span class="go">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">])</span>
<span class="go">gene C</span>
</pre></div>
</div>
<p>See <a class="reference internal" href="genomic.html#HTSeq.GenomicArray" title="HTSeq.GenomicArray"><code class="xref py py-class docutils literal"><span class="pre">GenomicArray</span></code></a> and <a class="reference internal" href="genomic.html#HTSeq.GenomicArrayOfSets" title="HTSeq.GenomicArrayOfSets"><code class="xref py py-class docutils literal"><span class="pre">GenomicArrayOfSets</span></code></a> for more sophisticated use.</p>
<div class="section" id="gff-reader-and-genomicfeature">
<h2><code class="docutils literal"><span class="pre">GFF_Reader</span></code> and <code class="docutils literal"><span class="pre">GenomicFeature</span></code><a class="headerlink" href="#gff-reader-and-genomicfeature" title="Permalink to this headline">¶</a></h2>
<p>One of the most common format for annotation data is <a class="reference external" href="http://www.sanger.ac.uk/resources/software/gff/spec.html">GFF</a> (which includes <a class="reference external" href="http://mblab.wustl.edu/GTF22.html">GTF</a> as
a sub-type). Hence, a parse for GFF files is included in HTSeq.</p>
<p>As usual, there is a parser class, called <strong>GFF_Reader</strong>, that can generate an
iterator of objects describing the features. These objects are of type :class`GenomicFeature`
and each describes one line of a GFF file. See Section <a class="reference internal" href="tour.html#tour"><span class="std std-ref">A tour through HTSeq</span></a> for an example.</p>
<dl class="class">
<dt id="HTSeq.GFF_Reader">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GFF_Reader</code><span class="sig-paren">(</span><em>filename_or_sequence</em>, <em>end_included=True</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GFF_Reader" title="Permalink to this definition">¶</a></dt>
<dd><p>As a subclass of <a class="reference internal" href="misc.html#HTSeq.FileOrSequence" title="HTSeq.FileOrSequence"><code class="xref py py-class docutils literal"><span class="pre">FileOrSequence</span></code></a>, GFF_Reader can be initialized either
with a file name or with an open file or another sequence of lines.</p>
<p>When requesting an iterator, it generates objects of type <a class="reference internal" href="#HTSeq.GenomicFeature" title="HTSeq.GenomicFeature"><code class="xref py py-class docutils literal"><span class="pre">GenomicFeature</span></code></a>.</p>
<p>The GFF specification is unclear on whether the end coordinate marks the last
base-pair of the feature (closed intervals, <code class="docutils literal"><span class="pre">end_included=True</span></code>) or the one
after (half-open intervals, <code class="docutils literal"><span class="pre">end_included=False</span></code>). The default, True, is
correct for Ensembl GTF files. If in doubt, look for a CDS or stop_codon
feature in you GFF file. Its length should be divisible by 3. If &#8220;end-start&#8221;
is divisible by 3, you need <code class="docutils literal"><span class="pre">end_included=False</span></code>. If &#8220;end-start+1&#8221; is
divisible by 3, you need <code class="docutils literal"><span class="pre">end_included=True</span></code>.</p>
<p>GFF_Reader will convert the coordinates from GFF standard (1-based, end
maybe included) to HTSeq standard (0-base, end not included) by subtracting
1 from the start position, and, for <code class="docutils literal"><span class="pre">end_included=True</span></code>, also subtract 1 from
the end position.</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.GFF_Reader.metadata">
<code class="descname">metadata</code><a class="headerlink" href="#HTSeq.GFF_Reader.metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>GFF_Reader skips all lines starting with a single &#8216;#&#8217; as this marks
a comment. However, lines starying with &#8216;##&#8217; contain meta data (at least
accoring to the Sanger Institute&#8217;s version of the GFF standard.) Such meta
data has the format <code class="docutils literal"><span class="pre">##key</span> <span class="pre">value</span></code>. When a metadata line is encountered,
it is added to the <code class="docutils literal"><span class="pre">metadata</span></code> dictionary.</p>
</dd></dl>

</div></blockquote>
</dd></dl>

<dl class="class">
<dt id="HTSeq.GenomicFeature">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GenomicFeature</code><span class="sig-paren">(</span><em>name</em>, <em>type_</em>, <em>interval</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicFeature" title="Permalink to this definition">¶</a></dt>
<dd><p>A GenomicFeature object always contains the following attributes:</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.GenomicFeature.name">
<code class="descname">name</code><a class="headerlink" href="#HTSeq.GenomicFeature.name" title="Permalink to this definition">¶</a></dt>
<dd><p>A name of ID for the feature. As the GFF format does not have a dedicated
field for this, the value of the first attribute in the <em>attributes</em> column is
assumed to be the name of ID.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicFeature.type">
<code class="descname">type</code><a class="headerlink" href="#HTSeq.GenomicFeature.type" title="Permalink to this definition">¶</a></dt>
<dd><p>The type of the feature, i.e., a string like <code class="docutils literal"><span class="pre">&quot;exon&quot;</span></code> or <code class="docutils literal"><span class="pre">&quot;gene&quot;</span></code>. For GFF
files, the 3rd column (<em>feature</em>) is taken as the type.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicFeature.interval">
<code class="descname">interval</code><a class="headerlink" href="#HTSeq.GenomicFeature.interval" title="Permalink to this definition">¶</a></dt>
<dd><p>The interval that the feature covers on the genome. For GFF files, this information is taken
from the first (<em>seqname</em>), the forth (<em>start</em>), the fifth (<em>end</em>), and the seventh (<em>strand</em>)
column.</p>
</dd></dl>

</div></blockquote>
<p>When created by a <a class="reference internal" href="#HTSeq.GFF_Reader" title="HTSeq.GFF_Reader"><code class="xref py py-class docutils literal"><span class="pre">GFF_Reader</span></code></a> object, the following attributes are also present, with the information
from the remaining GFF columns:</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.GenomicFeature.source">
<code class="descname">source</code><a class="headerlink" href="#HTSeq.GenomicFeature.source" title="Permalink to this definition">¶</a></dt>
<dd><p>The 2nd column, denoted <em>source</em> in the specification, and intended to specify the
data source.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicFeature.frame">
<code class="descname">frame</code><a class="headerlink" href="#HTSeq.GenomicFeature.frame" title="Permalink to this definition">¶</a></dt>
<dd><p>The 8th column (<em>frame</em>), giving the reading frame in case of a coding feature. Its value
is an integer (0, 1, or 2), or the string <code class="docutils literal"><span class="pre">'.'</span></code> in case that a frame is not specified or would not make sense.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicFeature.score">
<code class="descname">score</code><a class="headerlink" href="#HTSeq.GenomicFeature.score" title="Permalink to this definition">¶</a></dt>
<dd><p>The 6th column (<em>score</em>), giving some numerical score for the feature. Its value
is a float, or <code class="docutils literal"><span class="pre">'.'</span></code> in case that a score is not specified or would not make sense</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicFeature.attr">
<code class="descname">attr</code><a class="headerlink" href="#HTSeq.GenomicFeature.attr" title="Permalink to this definition">¶</a></dt>
<dd><p>The last (9th) column of a GFF file contains <em>attributes</em>, i.e. a list of name/value pairs.
These are transformed into a dict, such that, e.g., <code class="docutils literal"><span class="pre">gf.attr['gene_id']</span></code> gives the value
of the attribute <code class="docutils literal"><span class="pre">gene_id</span></code> in the feature described by <code class="docutils literal"><span class="pre">GenomicFeature</span></code> object <code class="docutils literal"><span class="pre">gf</span></code>.
The parser for the attribute field is reasonably flexible to deal with format variations
(it was never clearly established whetehr name and value should be sperarated by a colon or an
equal sign, and whether quotes need to be used) and also does a URL style decoding, as is often
required.</p>
</dd></dl>

</div></blockquote>
<p>In order to write a GFF file from a sequence of features, this method is provided:</p>
<blockquote>
<div><dl class="method">
<dt id="HTSeq.GenomicFeature.get_gff_line">
<code class="descname">get_gff_line</code><span class="sig-paren">(</span><em>with_equal_sign=False</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicFeature.get_gff_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a line to describe the feature in the GFF format. This works even if the optional
attributes given above are missing. Call this for each of your <code class="docutils literal"><span class="pre">GenomicFeature</span></code> objects
and write the lines into a file to get a GFF file.</p>
</dd></dl>

</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="HTSeq.parse_GFF_attribute_string">
<code class="descclassname">HTSeq.</code><code class="descname">parse_GFF_attribute_string</code><span class="sig-paren">(</span><em>attrStr</em>, <em>extra_return_first_value=False</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.parse_GFF_attribute_string" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the function that <a class="reference internal" href="#HTSeq.GFF_Reader" title="HTSeq.GFF_Reader"><code class="xref py py-class docutils literal"><span class="pre">GFF_Reader</span></code></a> uses to parse the attribute column. (See <a class="reference internal" href="#HTSeq.GenomicFeature.attr" title="HTSeq.GenomicFeature.attr"><code class="xref py py-attr docutils literal"><span class="pre">GenomicFeature.attr</span></code></a>.)
It returns a dict, or, if requested, a pair of the dict and the first value.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Features</a><ul>
<li><a class="reference internal" href="#gff-reader-and-genomicfeature"><code class="docutils literal"><span class="pre">GFF_Reader</span></code> and <code class="docutils literal"><span class="pre">GenomicFeature</span></code></a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="alignments.html"
                        title="previous chapter">Read alignments</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="otherparsers.html"
                        title="next chapter">Other parsers</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/features.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="otherparsers.html" title="Other parsers"
             >next</a> |</li>
        <li class="right" >
          <a href="alignments.html" title="Read alignments"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>