
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Index &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="#" />
    <link rel="search" title="Search" href="search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#Symbols"><strong>Symbols</strong></a>
 | <a href="#_"><strong>_</strong></a>
 | <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#Q"><strong>Q</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#V"><strong>V</strong></a>
 | <a href="#W"><strong>W</strong></a>
 | <a href="#X"><strong>X</strong></a>
 
</div>
<h2 id="Symbols">Symbols</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    -a &lt;minaqual&gt;, --a=&lt;minaqual&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-a">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -f &lt;format&gt;, --format=&lt;format&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-f">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -g &lt;gamma&gt;, --gamma=&lt;gamma&gt;

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-g">htseq-qa command line option</a>
</li>
      </ul></li>
      <li>
    -h, --help

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-h">htseq-count command line option</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-h">htseq-qa command line option</a>
</li>
      </ul></li>
      <li>
    -i &lt;id attribute&gt;, --idattr=&lt;id attribute&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-i">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -m &lt;mode&gt;, --mode=&lt;mode&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-m">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -m, --maxqual

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-m">htseq-qa command line option</a>
</li>
      </ul></li>
      <li>
    -n, --nosplit

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-n">htseq-qa command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    -o &lt;outfile&gt;, --outfile=&lt;outfile&gt;

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-o">htseq-qa command line option</a>
</li>
      </ul></li>
      <li>
    -o &lt;samout&gt;, --samout=&lt;samout&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-o">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -q, --quiet

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-q">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -r &lt;order&gt;, --order=&lt;order&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-r">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -r &lt;readlen&gt;, --readlength=&lt;readlen&gt;

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-r">htseq-qa command line option</a>
</li>
      </ul></li>
      <li>
    -s &lt;yes/no/reverse&gt;, --stranded=&lt;yes/no/reverse&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-s">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -t &lt;feature type&gt;, --type=&lt;feature type&gt;

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-t">htseq-count command line option</a>
</li>
      </ul></li>
      <li>
    -t &lt;type&gt;, --type=&lt;type&gt;

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-t">htseq-qa command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="_">_</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="misc.html#HTSeq.__version__">__version__ (in module HTSeq)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicArray.add_chrom">add_chrom() (HTSeq.GenomicArray method)</a>
</li>
      <li><a href="alignments.html#HTSeq.Alignment.aligned">aligned (HTSeq.Alignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.Alignment">Alignment (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.AlignmentWithSequenceReversal">AlignmentWithSequenceReversal (class in HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VariantCall.alt">alt (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.ChromVector.apply">apply() (HTSeq.ChromVector method)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.aQual">aQual (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="features.html#HTSeq.GenomicFeature.attr">attr (HTSeq.GenomicFeature attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicArray.auto_add_chroms">auto_add_chroms (HTSeq.GenomicArray attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.BAM_Reader">BAM_Reader (class in HTSeq)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.BED_Reader">BED_Reader (class in HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.BowtieAlignment">BowtieAlignment (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.BowtieReader">BowtieReader (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.bundle_multiple_alignments">bundle_multiple_alignments() (in module HTSeq)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.CigarOperation.check">check() (HTSeq.CigarOperation method)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.chrom">chrom (HTSeq.GenomicInterval attribute)</a>

      <ul>
        <li><a href="otherparsers.html#HTSeq.VariantCall.chrom">(HTSeq.VariantCall attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicArray.chrom_vectors">chrom_vectors (HTSeq.GenomicArray attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.cigar">cigar (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation">CigarOperation (class in HTSeq)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.contains">contains() (HTSeq.GenomicInterval method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="sequences.html#HTSeq.Sequence.descr">descr (HTSeq.Sequence attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicInterval.end">end (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.end_as_pos">end_as_pos (HTSeq.GenomicInterval attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicInterval.end_d">end_d (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.end_d_as_pos">end_d_as_pos (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.extend_to_include">extend_to_include() (HTSeq.GenomicInterval method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.failed_platform_qc">failed_platform_qc (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="sequences.html#HTSeq.FastaReader">FastaReader (class in HTSeq)</a>
</li>
      <li><a href="sequences.html#HTSeq.FastqReader">FastqReader (class in HTSeq)</a>
</li>
      <li><a href="misc.html#HTSeq.FileOrSequence">FileOrSequence (class in HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VariantCall.filter">filter (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VariantCall.format">format (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="misc.html#HTSeq.FileOrSequence.fos">fos (HTSeq.FileOrSequence attribute)</a>
</li>
      <li><a href="features.html#HTSeq.GenomicFeature.frame">frame (HTSeq.GenomicFeature attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicArray">GenomicArray (class in HTSeq)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicArrayOfSets">GenomicArrayOfSets (class in HTSeq)</a>
</li>
      <li><a href="features.html#HTSeq.GenomicFeature">GenomicFeature (class in HTSeq)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval">GenomicInterval (class in HTSeq)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval_from_directional">GenomicInterval_from_directional() (in module HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicPosition">GenomicPosition (class in HTSeq)</a>
</li>
      <li><a href="features.html#HTSeq.GenomicFeature.get_gff_line">get_gff_line() (HTSeq.GenomicFeature method)</a>
</li>
      <li><a href="misc.html#HTSeq.FileOrSequence.get_line_number_string">get_line_number_string() (HTSeq.FileOrSequence method)</a>
</li>
      <li><a href="sequences.html#HTSeq.Sequence.get_reverse_complement">get_reverse_complement() (HTSeq.Sequence method)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.get_sam_line">get_sam_line() (HTSeq.SAM_Alignment method)</a>
</li>
      <li><a href="features.html#HTSeq.GFF_Reader">GFF_Reader (class in HTSeq)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    htseq-count command line option

      <ul>
        <li><a href="count.html#cmdoption-htseq-count-a">-a &lt;minaqual&gt;, --a=&lt;minaqual&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-f">-f &lt;format&gt;, --format=&lt;format&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-h">-h, --help</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-i">-i &lt;id attribute&gt;, --idattr=&lt;id attribute&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-m">-m &lt;mode&gt;, --mode=&lt;mode&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-o">-o &lt;samout&gt;, --samout=&lt;samout&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-q">-q, --quiet</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-r">-r &lt;order&gt;, --order=&lt;order&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-s">-s &lt;yes/no/reverse&gt;, --stranded=&lt;yes/no/reverse&gt;</a>
</li>
        <li><a href="count.html#cmdoption-htseq-count-t">-t &lt;feature type&gt;, --type=&lt;feature type&gt;</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    htseq-qa command line option

      <ul>
        <li><a href="qa.html#cmdoption-htseq-qa-g">-g &lt;gamma&gt;, --gamma=&lt;gamma&gt;</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-h">-h, --help</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-m">-m, --maxqual</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-n">-n, --nosplit</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-o">-o &lt;outfile&gt;, --outfile=&lt;outfile&gt;</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-r">-r &lt;readlen&gt;, --readlength=&lt;readlen&gt;</a>
</li>
        <li><a href="qa.html#cmdoption-htseq-qa-t">-t &lt;type&gt;, --type=&lt;type&gt;</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VariantCall.id">id (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.inferred_insert_size">inferred_insert_size (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VariantCall.info">info (HTSeq.VariantCall attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="features.html#HTSeq.GenomicFeature.interval">interval (HTSeq.GenomicFeature attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.is_contained_in">is_contained_in() (HTSeq.GenomicInterval method)</a>
</li>
      <li><a href="alignments.html#HTSeq.Alignment.iv">iv (HTSeq.Alignment attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicInterval.length">length (HTSeq.GenomicInterval attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="misc.html#HTSeq.FileOrSequence.line_no">line_no (HTSeq.FileOrSequence attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.mate_aligned">mate_aligned (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.mate_start">mate_start (HTSeq.SAM_Alignment attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="features.html#HTSeq.GFF_Reader.metadata">metadata (HTSeq.GFF_Reader attribute)</a>

      <ul>
        <li><a href="otherparsers.html#HTSeq.VCF_Reader.metadata">(HTSeq.VCF_Reader attribute)</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="features.html#HTSeq.GenomicFeature.name">name (HTSeq.GenomicFeature attribute)</a>

      <ul>
        <li><a href="sequences.html#HTSeq.Sequence.name">(HTSeq.Sequence attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SolexaExportAlignment.nomatch_code">nomatch_code (HTSeq.SolexaExportAlignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.nor_primary_alignment">nor_primary_alignment (HTSeq.SAM_Alignment attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.optional_field">optional_field (HTSeq.SAM_Alignment attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.optional_fields">optional_fields (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.overlaps">overlaps() (HTSeq.GenomicInterval method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.pair_SAM_alignments">pair_SAM_alignments() (in module HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.pair_SAM_alignments_with_buffer">pair_SAM_alignments_with_buffer() (in module HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.Alignment.paired_end">paired_end (HTSeq.Alignment attribute)</a>
</li>
      <li><a href="features.html#HTSeq.parse_GFF_attribute_string">parse_GFF_attribute_string() (in module HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.SolexaExportAlignment.passed_filter">passed_filter (HTSeq.SolexaExportAlignment attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.pcr_or_optical_duplicate">pcr_or_optical_duplicate (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.pe_which">pe_which (HTSeq.SAM_Alignment attribute)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VariantCall.pos">pos (HTSeq.VariantCall attribute)</a>

      <ul>
        <li><a href="genomic.html#HTSeq.pos">(in module HTSeq)</a>
</li>
      </ul></li>
      <li><a href="alignments.html#HTSeq.SAM_Alignment.proper_pair">proper_pair (HTSeq.SAM_Alignment attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="Q">Q</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="sequences.html#HTSeq.SequenceWithQuality.qual">qual (HTSeq.SequenceWithQuality attribute)</a>

      <ul>
        <li><a href="otherparsers.html#HTSeq.VariantCall.qual">(HTSeq.VariantCall attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="sequences.html#HTSeq.SequenceWithQuality.qualstr">qualstr (HTSeq.SequenceWithQuality attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation.query_from">query_from (HTSeq.CigarOperation attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation.query_to">query_to (HTSeq.CigarOperation attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.Alignment.read">read (HTSeq.Alignment attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.AlignmentWithSequenceReversal.read_as_aligned">read_as_aligned (HTSeq.AlignmentWithSequenceReversal attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.AlignmentWithSequenceReversal.read_as_sequenced">read_as_sequenced (HTSeq.AlignmentWithSequenceReversal attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VariantCall.ref">ref (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation.ref_iv">ref_iv (HTSeq.CigarOperation attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.BowtieAlignment.reserved">reserved (HTSeq.BowtieAlignment attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="alignments.html#HTSeq.SAM_Alignment">SAM_Alignment (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.SAM_Reader">SAM_Reader (class in HTSeq)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VariantCall.samples">samples (HTSeq.VariantCall attribute)</a>
</li>
      <li><a href="features.html#HTSeq.GenomicFeature.score">score (HTSeq.GenomicFeature attribute)</a>
</li>
      <li><a href="sequences.html#HTSeq.Sequence.seq">seq (HTSeq.Sequence attribute)</a>
</li>
      <li><a href="sequences.html#HTSeq.Sequence">Sequence (class in HTSeq)</a>
</li>
      <li><a href="sequences.html#HTSeq.SequenceWithQualities">SequenceWithQualities (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation.size">size (HTSeq.CigarOperation attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.SolexaExportAlignment">SolexaExportAlignment (class in HTSeq)</a>
</li>
      <li><a href="alignments.html#HTSeq.SolexaExportReader">SolexaExportReader (class in HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="features.html#HTSeq.GenomicFeature.source">source (HTSeq.GenomicFeature attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.start">start (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.start_as_pos">start_as_pos (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.start_d">start_d (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.start_d_as_pos">start_d_as_pos (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.ChromVector.steps">steps() (HTSeq.ChromVector method)</a>

      <ul>
        <li><a href="genomic.html#HTSeq.GenomicArray.steps">(HTSeq.GenomicArray method)</a>
</li>
      </ul></li>
      <li><a href="genomic.html#HTSeq.GenomicInterval.strand">strand (HTSeq.GenomicInterval attribute)</a>
</li>
      <li><a href="genomic.html#HTSeq.GenomicArray.stranded">stranded (HTSeq.GenomicArray attribute)</a>
</li>
      <li><a href="alignments.html#HTSeq.BowtieAlignment.substitutions">substitutions (HTSeq.BowtieAlignment attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="sequences.html#HTSeq.Sequence.trim_left_end">trim_left_end() (HTSeq.Sequence method)</a>
</li>
      <li><a href="sequences.html#HTSeq.SequenceWithQualities.trim_left_end_with_quals">trim_left_end_with_quals() (HTSeq.SequenceWithQualities method)</a>
</li>
      <li><a href="sequences.html#HTSeq.Sequence.trim_right_end">trim_right_end() (HTSeq.Sequence method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="sequences.html#HTSeq.SequenceWithQualities.trim_right_end_with_quals">trim_right_end_with_quals() (HTSeq.SequenceWithQualities method)</a>
</li>
      <li><a href="alignments.html#HTSeq.CigarOperation.type">type (HTSeq.CigarOperation attribute)</a>

      <ul>
        <li><a href="features.html#HTSeq.GenomicFeature.type">(HTSeq.GenomicFeature attribute)</a>
</li>
      </ul></li>
      <li><a href="genomic.html#HTSeq.GenomicArray.typecode">typecode (HTSeq.GenomicArray attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VariantCall">VariantCall (class in HTSeq)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VariantCall.unpack_info">VariantCall.unpack_info() (in module HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.VCF_Reader">VCF_Reader (class in HTSeq)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VCF_Reader.make_info_dict">VCF_Reader.make_info_dict() (in module HTSeq)</a>
</li>
      <li><a href="otherparsers.html#HTSeq.VCF_Reader.parse_meta">VCF_Reader.parse_meta() (in module HTSeq)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="W">W</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="otherparsers.html#HTSeq.WiggleReader">WiggleReader (class in HTSeq)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicArray.write_bedgraph_file">write_bedgraph_file() (HTSeq.GenomicArray method)</a>
</li>
      <li><a href="sequences.html#HTSeq.SequenceWithQuality.write_to_fastq_file">write_to_fastq_file() (HTSeq.SequenceWithQuality method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="X">X</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicInterval.xrange">xrange() (HTSeq.GenomicInterval method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="genomic.html#HTSeq.GenomicInterval.xrange_d">xrange_d() (HTSeq.GenomicInterval method)</a>
</li>
  </ul></td>
</tr></table>



          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

   

<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>