<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Genomic intervals and genomic arrays &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Read alignments" href="alignments.html" />
    <link rel="prev" title="Sequences and FASTA/FASTQ files" href="sequences.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="alignments.html" title="Read alignments"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="sequences.html" title="Sequences and FASTA/FASTQ files"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="genomic-intervals-and-genomic-arrays">
<span id="genomic"></span><h1>Genomic intervals and genomic arrays<a class="headerlink" href="#genomic-intervals-and-genomic-arrays" title="Permalink to this headline">¶</a></h1>
<div class="section" id="genomicinterval">
<h2><code class="docutils literal"><span class="pre">GenomicInterval</span></code><a class="headerlink" href="#genomicinterval" title="Permalink to this headline">¶</a></h2>
<p>A genomic interval is a consecutive stretch on a genomic sequence such as a chromosome.
It is represented by a <code class="docutils literal"><span class="pre">GenomicInterval</span></code> object.</p>
<dl class="docutils">
<dt>Instantiation</dt>
<dd><dl class="first last class">
<dt id="HTSeq.GenomicInterval">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GenomicInterval</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>end</em>, <em>strand</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt><code class="docutils literal"><span class="pre">chrom</span></code> (string)</dt>
<dd>The name of a sequence (i.e., chromosome, contig, or the like).</dd>
<dt><code class="docutils literal"><span class="pre">start</span></code>  (int)</dt>
<dd>The start of the interval. Even on the reverse strand,
this is always the smaller of the two values &#8216;start&#8217; and &#8216;end&#8217;.
Note that all positions should be given and interpreted as 0-based value!</dd>
<dt><code class="docutils literal"><span class="pre">end</span></code>  (int)</dt>
<dd>The end of the interval. Following Python convention for
ranges, this in one more than the coordinate of the last base
that is considered part of the sequence.</dd>
<dt><code class="docutils literal"><span class="pre">strand</span></code>  (string)</dt>
<dd>The strand, as a single character, <code class="docutils literal"><span class="pre">'+'</span></code>, <code class="docutils literal"><span class="pre">'-'</span></code>, or <code class="docutils literal"><span class="pre">'.'</span></code>.
<code class="docutils literal"><span class="pre">'.'</span></code> indicates that the strand is irrelevant.</dd>
</dl>
</dd></dl>

</dd>
<dt>Representation and string conversion</dt>
<dd><p class="first">The class&#8217;s <code class="docutils literal"><span class="pre">__str__</span></code> method gives a spcae-saving description of the
interval, the <code class="docutils literal"><span class="pre">__repr__</span></code> method is a bit more verbose:</p>
<div class="last highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">iv</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr3&quot;</span><span class="p">,</span> <span class="mi">123203</span><span class="p">,</span> <span class="mi">127245</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">iv</span><span class="p">)</span>
<span class="go">chr3:[123203,127245)/+</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">iv</span>
<span class="go">&lt;GenomicInterval object &#39;chr3&#39;, [123203,127245), strand &#39;+&#39;&gt;</span>
</pre></div>
</div>
</dd>
</dl>
<p>Attributes</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.GenomicInterval.chrom">
<code class="descclassname">GenomicInterval.</code><code class="descname">chrom</code><a class="headerlink" href="#HTSeq.GenomicInterval.chrom" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.start">
<code class="descclassname">GenomicInterval.</code><code class="descname">start</code><a class="headerlink" href="#HTSeq.GenomicInterval.start" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.end">
<code class="descclassname">GenomicInterval.</code><code class="descname">end</code><a class="headerlink" href="#HTSeq.GenomicInterval.end" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.strand">
<code class="descclassname">GenomicInterval.</code><code class="descname">strand</code><a class="headerlink" href="#HTSeq.GenomicInterval.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>as above</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicInterval.start_d">
<code class="descclassname">GenomicInterval.</code><code class="descname">start_d</code><a class="headerlink" href="#HTSeq.GenomicInterval.start_d" title="Permalink to this definition">¶</a></dt>
<dd><p>The &#8220;directional start&#8221; position. This is the position of the
first base of the interval, taking the strand into account. Hence,
this is the same as <code class="docutils literal"><span class="pre">start</span></code> except when <code class="docutils literal"><span class="pre">strand</span> <span class="pre">==</span> <span class="pre">'-'</span></code>, in which
case it is <code class="docutils literal"><span class="pre">end-1</span></code>.</p>
<p>Note that if you set <code class="docutils literal"><span class="pre">start_d</span></code>, both <code class="docutils literal"><span class="pre">start</span></code> and <code class="docutils literal"><span class="pre">end</span></code> are changed,
such that the interval gets the requested new directional start and its
length stays unchanged.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicInterval.end_d">
<code class="descclassname">GenomicInterval.</code><code class="descname">end_d</code><a class="headerlink" href="#HTSeq.GenomicInterval.end_d" title="Permalink to this definition">¶</a></dt>
<dd><p>The &#8220;directional end&#8221;: The same as <code class="docutils literal"><span class="pre">end</span></code>, unless <code class="docutils literal"><span class="pre">strand=='-'</span></code>,
in which case it is <code class="docutils literal"><span class="pre">start-1</span></code>. This convention allows to go from
<code class="docutils literal"><span class="pre">start_d</span></code> to <code class="docutils literal"><span class="pre">end_d</span></code> (not including, as usual in Python, the last
value) and get all bases in &#8220;reading&#8221; direction.</p>
<p><code class="docutils literal"><span class="pre">end_d</span></code> is not writable.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicInterval.length">
<code class="descclassname">GenomicInterval.</code><code class="descname">length</code><a class="headerlink" href="#HTSeq.GenomicInterval.length" title="Permalink to this definition">¶</a></dt>
<dd><p>The length is calculated as end - start. If you set the length,
<code class="docutils literal"><span class="pre">start_d</span></code> will be preserved, i.e., <code class="docutils literal"><span class="pre">end</span></code> is changed,
unless the strand is <code class="docutils literal"><span class="pre">-</span></code>, in which case <code class="docutils literal"><span class="pre">start</span></code> is changed.</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicInterval.start_as_pos">
<code class="descclassname">GenomicInterval.</code><code class="descname">start_as_pos</code><a class="headerlink" href="#HTSeq.GenomicInterval.start_as_pos" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.end_as_pos">
<code class="descclassname">GenomicInterval.</code><code class="descname">end_as_pos</code><a class="headerlink" href="#HTSeq.GenomicInterval.end_as_pos" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.start_d_as_pos">
<code class="descclassname">GenomicInterval.</code><code class="descname">start_d_as_pos</code><a class="headerlink" href="#HTSeq.GenomicInterval.start_d_as_pos" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.end_d_as_pos">
<code class="descclassname">GenomicInterval.</code><code class="descname">end_d_as_pos</code><a class="headerlink" href="#HTSeq.GenomicInterval.end_d_as_pos" title="Permalink to this definition">¶</a></dt>
<dd><p>These attributes return <a class="reference internal" href="#HTSeq.GenomicPosition" title="HTSeq.GenomicPosition"><code class="xref py py-class docutils literal"><span class="pre">GenomicPosition</span></code></a> objects referring to the
respective positions.</p>
</dd></dl>

</div></blockquote>
<dl class="docutils">
<dt>Directional instantiation</dt>
<dd><dl class="first last function">
<dt id="HTSeq.GenomicInterval_from_directional">
<code class="descclassname">HTSeq.</code><code class="descname">GenomicInterval_from_directional</code><span class="sig-paren">(</span><em>chrom</em>, <em>start_d</em>, <em>length</em>, <em>strand=&quot;.&quot;</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval_from_directional" title="Permalink to this definition">¶</a></dt>
<dd><p>This function allows to create a new <code class="docutils literal"><span class="pre">GenomicInterval</span></code> object specifying
directional start and length instead of start and end.</p>
</dd></dl>

</dd>
<dt>Methods</dt>
<dd><dl class="first method">
<dt id="HTSeq.GenomicInterval.is_contained_in">
<code class="descclassname">GenomicInterval.</code><code class="descname">is_contained_in</code><span class="sig-paren">(</span><em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.is_contained_in" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.contains">
<code class="descclassname">GenomicInterval.</code><code class="descname">contains</code><span class="sig-paren">(</span><em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.contains" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.overlaps">
<code class="descclassname">GenomicInterval.</code><code class="descname">overlaps</code><span class="sig-paren">(</span><em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>These methods test whether the object is contained in, contains, or overlaps
the second <code class="docutils literal"><span class="pre">GenomicInterval</span></code> object <code class="docutils literal"><span class="pre">iv</span></code>.</p>
<p>For any of of these conditions
to be true, the <code class="docutils literal"><span class="pre">start</span></code> and <code class="docutils literal"><span class="pre">end</span></code> values have to be appropriate, and furthermore,
the <code class="docutils literal"><span class="pre">chrom</span></code> values have to be equal and the <code class="docutils literal"><span class="pre">strand</span></code> values consistent. The latter
means that the strands have to be the same if both intervals have strand
information. However, if at least one of the objects has <code class="docutils literal"><span class="pre">strand</span> <span class="pre">==</span> <span class="pre">'.'</span></code>,
the strand information of the other object is disregarded.</p>
<p>Note that all three methods return <code class="docutils literal"><span class="pre">True</span></code> for identical intervals.</p>
</dd></dl>

<dl class="method">
<dt id="HTSeq.GenomicInterval.xrange">
<code class="descclassname">GenomicInterval.</code><code class="descname">xrange</code><span class="sig-paren">(</span><em>step = 1</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.xrange" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicInterval.xrange_d">
<code class="descclassname">GenomicInterval.</code><code class="descname">xrange_d</code><span class="sig-paren">(</span><em>step = 1</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.xrange_d" title="Permalink to this definition">¶</a></dt>
<dd><p>These methods yield iterators of :class:GenomicPosition objects from
<code class="docutils literal"><span class="pre">start</span></code> to <code class="docutils literal"><span class="pre">end</span></code> (or, for <code class="docutils literal"><span class="pre">xrange_d</span></code> from <code class="docutils literal"><span class="pre">start_d</span></code> to <code class="docutils literal"><span class="pre">end_d</span></code>).</p>
</dd></dl>

<dl class="last method">
<dt id="HTSeq.GenomicInterval.extend_to_include">
<code class="descclassname">GenomicInterval.</code><code class="descname">extend_to_include</code><span class="sig-paren">(</span><em>iv</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicInterval.extend_to_include" title="Permalink to this definition">¶</a></dt>
<dd><p>Change the object&#8217;s <code class="docutils literal"><span class="pre">start</span></code> end <code class="docutils literal"><span class="pre">end</span></code> values such that <code class="docutils literal"><span class="pre">iv</span></code> becomes contained.</p>
</dd></dl>

</dd>
</dl>
<p>Special methods</p>
<blockquote>
<div><p><code class="docutils literal"><span class="pre">GenomicInterval</span></code> implements the methods necessary for</p>
<ul class="simple">
<li>obtaining a copy of the object (the <code class="docutils literal"><span class="pre">copy</span></code> method)</li>
<li>pickling the object</li>
<li>representing the object and converting it to a string (see above)</li>
<li>comparing two GenomicIntervals for equality and inequality</li>
<li>hashing the object</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="genomicposition">
<h2><code class="docutils literal"><span class="pre">GenomicPosition</span></code><a class="headerlink" href="#genomicposition" title="Permalink to this headline">¶</a></h2>
<p>A <code class="docutils literal"><span class="pre">GenomicPosition</span></code> represents the position of a single base or base pair, i.e., it is
an interval of length 1, and hence, the class is a subclass of :class:GenomicInterval.</p>
<dl class="class">
<dt id="HTSeq.GenomicPosition">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GenomicPosition</code><span class="sig-paren">(</span><em>chrom</em>, <em>pos</em>, <em>strand='.'</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicPosition" title="Permalink to this definition">¶</a></dt>
<dd><p>The initialisation is as for a :class:GenomicInterval object, but no <code class="docutils literal"><span class="pre">length</span></code> argument is passed.</p>
</dd></dl>

<p>Attributes</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.pos">
<code class="descclassname">HTSeq.</code><code class="descname">pos</code><a class="headerlink" href="#HTSeq.pos" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>pos</strong> is an alias for <a class="reference internal" href="#HTSeq.GenomicInterval.start_d" title="HTSeq.GenomicInterval.start_d"><code class="xref py py-attr docutils literal"><span class="pre">GenomicInterval.start_d</span></code></a>.</p>
</dd></dl>

<p>All other attributes of <a class="reference internal" href="#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> are still exposed. Refrain from
using them, unless you want to use the object as an interval, not as a position.
Some of them are now read-only to prevent the length to be changed.</p>
</div></blockquote>
</div>
<div class="section" id="genomicarray">
<h2><code class="docutils literal"><span class="pre">GenomicArray</span></code><a class="headerlink" href="#genomicarray" title="Permalink to this headline">¶</a></h2>
<p>A <code class="docutils literal"><span class="pre">GenomicArray</span></code> is a collection of <code class="xref py py-class docutils literal"><span class="pre">ChromVector</span></code> objects, either one or two
for each chromosome of a genome. It allows to access the data in these
transparently via <a class="reference internal" href="#HTSeq.GenomicInterval" title="HTSeq.GenomicInterval"><code class="xref py py-class docutils literal"><span class="pre">GenomicInterval</span></code></a> objects.</p>
<p><strong>Note: ``GenomicArray``&#8217;s interface changed significantly in version 0.5.0. Please
see the Version History page.</strong></p>
<dl class="docutils">
<dt>Instantiation</dt>
<dd><dl class="first class">
<dt id="HTSeq.GenomicArray">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GenomicArray</code><span class="sig-paren">(</span><em>chroms</em>, <em>stranded=True</em>, <em>typecode='d'</em>, <em>storage='step'</em>, <em>memmap_dir=''</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicArray" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>Creates a <code class="docutils literal"><span class="pre">GenomicArray</span></code>.</p>
<p>If <code class="docutils literal"><span class="pre">chroms</span></code> is a list of chromosome names,  two (or one, see below) <code class="docutils literal"><span class="pre">ChromVector</span></code>
objects for each chromosome are created, with start index 0 and indefinite
length. If <code class="docutils literal"><span class="pre">chroms</span></code> is a <code class="docutils literal"><span class="pre">dict</span></code>, the keys are used for the chromosome names
and the values should be the lengths of the chromosome, i.e., the ChromVectors
index ranges are then from 0 to these lengths. (Note that the term chromosome
is used only for convenience. Of course, you can as well specify contig IDs
or the like.) Finally, if <code class="docutils literal"><span class="pre">chroms</span></code> is the string <code class="docutils literal"><span class="pre">&quot;auto&quot;</span></code>, the GenomicArray
is created without any chromosomes but whenever the user attempts to assign a
value to a yet unknown chromosome, a new one is automatically created with
<a class="reference internal" href="#HTSeq.GenomicArray.add_chrom" title="HTSeq.GenomicArray.add_chrom"><code class="xref py py-meth docutils literal"><span class="pre">GenomicArray.add_chrom()</span></code></a>.</p>
<p>If <code class="docutils literal"><span class="pre">stranded</span></code> is <code class="docutils literal"><span class="pre">True</span></code>, two <code class="docutils literal"><span class="pre">StepVector</span></code> objects are created for each chromosome,
one for the &#8216;+&#8217; and one for the &#8216;-&#8216; strand. For <code class="docutils literal"><span class="pre">stranded</span> <span class="pre">==</span> <span class="pre">False</span></code>, only one
<code class="docutils literal"><span class="pre">StepVector</span></code> per chromosome is used. In that case, the strand argument of
all <code class="docutils literal"><span class="pre">GenomicInterval</span></code> objects that are used later to specify regions in the
<code class="docutils literal"><span class="pre">GenomicArray</span></code> are ignored.</p>
<p>The <code class="docutils literal"><span class="pre">typecode</span></code> determines the data type and is as in <code class="docutils literal"><span class="pre">numpy</span></code>, i.e.:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">'d'</span></code> for float values (C type &#8216;double&#8217;),</li>
<li><code class="docutils literal"><span class="pre">'i'</span></code> for int values,</li>
<li><code class="docutils literal"><span class="pre">'b'</span></code> for Boolean values,</li>
<li><code class="docutils literal"><span class="pre">'O'</span></code> for arbitrary Python objects as value.</li>
</ul>
</div></blockquote>
<p>The storage mode determines how the ChromVectors store the data internally:</p>
<blockquote class="last">
<div><ul class="simple">
<li>mode <code class="docutils literal"><span class="pre">'step'</span></code>: A step vector is used. This is the default and useful
for large amounts of data which may stay constant along a range of indices.
Each such step is stored internally as a node in a red-black tree.</li>
<li>mode <code class="docutils literal"><span class="pre">'ndarray'</span></code>: A 1D numpy array is used internally. This is useful
if the data changes a lot, and steps are hence inefficient. Using this mode
requires that chromosome lengths are specified.</li>
<li>mode <code class="docutils literal"><span class="pre">memmap</span></code>: This is useful for large amounts of data with very short
steps, where <code class="docutils literal"><span class="pre">step</span></code> is inefficient, but a numpy vectors would not fit
into memory. A numpy memmap is used that stores the whole vector in a
file on disk and transparently maps into memory windows of the data. This
mode requires chromosome lengths, and specification of a directory, via
the <code class="docutils literal"><span class="pre">memmap_dir</span></code> argument, to store the temporary files in. It is not
suitable for type code <code class="docutils literal"><span class="pre">O</span></code>.</li>
</ul>
</div></blockquote>
</dd>
</dl>
<p>Attributes</p>
<blockquote>
<div><dl class="attribute">
<dt id="HTSeq.GenomicArray.stranded">
<code class="descclassname">GenomicArray.</code><code class="descname">stranded</code><a class="headerlink" href="#HTSeq.GenomicArray.stranded" title="Permalink to this definition">¶</a></dt>
<dt id="HTSeq.GenomicArray.typecode">
<code class="descclassname">GenomicArray.</code><code class="descname">typecode</code><a class="headerlink" href="#HTSeq.GenomicArray.typecode" title="Permalink to this definition">¶</a></dt>
<dd><p>see above</p>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicArray.chrom_vectors">
<code class="descclassname">GenomicArray.</code><code class="descname">chrom_vectors</code><a class="headerlink" href="#HTSeq.GenomicArray.chrom_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>a dict of dicts of <code class="docutils literal"><span class="pre">ChromVector</span></code> objects, using the chromosome
names, and the strand as keys:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">..</span> <span class="n">doctest</span><span class="p">::</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">(</span> <span class="p">[</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span> <span class="p">],</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">ga</span><span class="o">.</span><span class="n">chrom_vectors</span><span class="o">.</span><span class="n">items</span><span class="p">())</span> 
<span class="go">[(&#39;chr1&#39;, {&#39;.&#39;: &lt;ChromVector object, chr1:[0,Inf)/., step&gt;}),</span>
<span class="go"> (&#39;chr2&#39;, {&#39;.&#39;: &lt;ChromVector object, chr2:[0,Inf)/., step&gt;})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">(</span> <span class="p">[</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span> <span class="p">],</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">True</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">([(</span><span class="n">st</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">st</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()))</span> <span class="k">for</span> <span class="n">st</span> <span class="ow">in</span> <span class="n">ga</span><span class="o">.</span><span class="n">chrom_vectors</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span>  
<span class="go">[(&#39;chr1&#39;, [(&#39;+&#39;, &lt;ChromVector object, chr1:[0,Inf)/+, step&gt;),</span>
<span class="go">           (&#39;-&#39;, &lt;ChromVector object, chr1:[0,Inf)/-, step&gt;)]),</span>
<span class="go"> (&#39;chr2&#39;, [(&#39;+&#39;, &lt;ChromVector object, chr2:[0,Inf)/+, step&gt;),</span>
<span class="go">           (&#39;-&#39;, &lt;ChromVector object, chr2:[0,Inf)/-, step&gt;)])]</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="HTSeq.GenomicArray.auto_add_chroms">
<code class="descclassname">GenomicArray.</code><code class="descname">auto_add_chroms</code><a class="headerlink" href="#HTSeq.GenomicArray.auto_add_chroms" title="Permalink to this definition">¶</a></dt>
<dd><p>A boolean. This attribute is set to True if the GenomicArray was created with the <code class="docutils literal"><span class="pre">&quot;auto&quot;</span></code>
arguments for the <code class="docutils literal"><span class="pre">chroms</span></code> parameter. If it is true, an new chromosome
will be added whenever needed.</p>
</dd></dl>

</div></blockquote>
<dl class="docutils">
<dt>Data access</dt>
<dd><p class="first">To access the data, use :class:GenomicInterval objects.</p>
<p>To set an single position or an interval, use:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span> <span class="o">=</span> <span class="mi">7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">400</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span> <span class="o">=</span> <span class="mi">20</span>
</pre></div>
</div>
<p>To read a single position:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span>
<span class="go">20.0</span>
</pre></div>
</div>
<dl class="method">
<dt id="HTSeq.ChromVector.steps">
<code class="descclassname">ChromVector.</code><code class="descname">steps</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.ChromVector.steps" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<p>To read an interval, use a <code class="docutils literal"><span class="pre">GenomicInterval</span></code> object as index, and
obtain a <code class="docutils literal"><span class="pre">ChromVector</span></code> with a sub-view:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">iv</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">450</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v</span> <span class="o">=</span> <span class="n">ga</span><span class="p">[</span> <span class="n">iv</span> <span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v</span>
<span class="go">&lt;ChromVector object, chr1:[250,450)/+, step&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span> <span class="n">v</span><span class="o">.</span><span class="n">steps</span><span class="p">()</span> <span class="p">)</span>  
<span class="go">[(&lt;GenomicInterval object &#39;chr1&#39;, [250,400), strand &#39;+&#39;&gt;, 20.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [400,450), strand &#39;+&#39;&gt;, 0.0)]</span>
</pre></div>
</div>
<p>Note that you get ( interval, value ) pairs , i.e., you can conveniently cycle
through them with:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">iv</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">ga</span><span class="p">[</span> <span class="n">iv</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="n">iv</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
<span class="go">chr1:[250,400)/+ 20.0</span>
<span class="go">chr1:[400,450)/+ 0.0</span>
</pre></div>
</div>
<dl class="last method">
<dt id="HTSeq.GenomicArray.steps">
<code class="descclassname">GenomicArray.</code><code class="descname">steps</code><span class="sig-paren">(</span><em>self</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicArray.steps" title="Permalink to this definition">¶</a></dt>
<dd><p>You can get all steps from all chromosomes by calling the arrays own <code class="docutils literal"><span class="pre">steps</span></code>
method.</p>
</dd></dl>

</dd>
</dl>
<p>Modifying values</p>
<blockquote>
<div><p>ChromVector implements the <code class="docutils literal"><span class="pre">__iadd__</span></code> method. Hence you can use <code class="docutils literal"><span class="pre">+=</span></code>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">290</span><span class="p">,</span> <span class="mi">310</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span> <span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">450</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">()</span> <span class="p">)</span>  
<span class="go">[(&lt;GenomicInterval object &#39;chr1&#39;, [250,290), strand &#39;+&#39;&gt;, 20.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [290,310), strand &#39;+&#39;&gt;, 1020.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [310,400), strand &#39;+&#39;&gt;, 20.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [400,450), strand &#39;+&#39;&gt;, 0.0)]</span>
</pre></div>
</div>
<p>To do manipulations other than additions, use Chromvector&#8217;s <code class="docutils literal"><span class="pre">apply</span></code> method:</p>
<dl class="method">
<dt id="HTSeq.ChromVector.apply">
<code class="descclassname">ChromVector.</code><code class="descname">apply</code><span class="sig-paren">(</span><em>func</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.ChromVector.apply" title="Permalink to this definition">¶</a></dt>
<dd><div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">290</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">*</span> <span class="mi">2</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span> <span class="n">ga</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">250</span><span class="p">,</span> <span class="mi">450</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">()</span> <span class="p">)</span> 
<span class="go">[(&lt;GenomicInterval object &#39;chr1&#39;, [250,290), strand &#39;+&#39;&gt;, 20.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [290,300), strand &#39;+&#39;&gt;, 2040.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [300,310), strand &#39;+&#39;&gt;, 1020.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [310,400), strand &#39;+&#39;&gt;, 20.0),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [400,450), strand &#39;+&#39;&gt;, 0.0)]</span>
</pre></div>
</div>
</dd></dl>

</div></blockquote>
<dl class="docutils">
<dt>Writing to a file</dt>
<dd><dl class="first last method">
<dt id="HTSeq.GenomicArray.write_bedgraph_file">
<code class="descclassname">GenomicArray.</code><code class="descname">write_bedgraph_file</code><span class="sig-paren">(</span><em>file_or_filename</em>, <em>strand=&quot;.&quot;</em>, <em>track_options=&quot;&quot;</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicArray.write_bedgraph_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out the data in the GenomicArray as a <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/bedgraph.html">BedGraph</a> track. This is a subtype of the <a class="reference external" href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">Wiggle</a> format
(i.e., the file extension is usually &#8221;.wig&#8221;) and such files can be conveniently viewed in
a genome browser, e.g., with <a class="reference external" href="http://igb.bioviz.org/">IGB</a>.</p>
<p>This works only for numerical data, i.e., <code class="docutils literal"><span class="pre">datatype</span></code> <code class="docutils literal"><span class="pre">'i'</span></code> or <code class="docutils literal"><span class="pre">'d'</span></code>. As a bedgraph track
cannot store strand information, you have to specify either <code class="docutils literal"><span class="pre">'+'</span></code> or <code class="docutils literal"><span class="pre">'-'</span></code> as the
strand argument if your <code class="docutils literal"><span class="pre">GenomicArray</span></code> is stranded (<code class="docutils literal"><span class="pre">stranded==True</span></code>). Typically, you
will write two wiggle files, one for each strand, and display them together.</p>
</dd></dl>

</dd>
<dt>Adding a chromosome</dt>
<dd><dl class="first last method">
<dt id="HTSeq.GenomicArray.add_chrom">
<code class="descclassname">GenomicArray.</code><code class="descname">add_chrom</code><span class="sig-paren">(</span><em>chrom</em>, <em>length=sys.maxint</em>, <em>start_index=0</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicArray.add_chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds step vector(s) for a further chromosome. This is useful if you do not have a full
list of chromosome names yet when instantiating the <code class="docutils literal"><span class="pre">GenomicArray</span></code>.</p>
</dd></dl>

</dd>
</dl>
</div>
<div class="section" id="genomicarrayofsets">
<h2><code class="docutils literal"><span class="pre">GenomicArrayOfSets</span></code><a class="headerlink" href="#genomicarrayofsets" title="Permalink to this headline">¶</a></h2>
<p>A <code class="docutils literal"><span class="pre">GenomicArrayOfSets</span></code> is a sub-class of <a class="reference internal" href="#HTSeq.GenomicArray" title="HTSeq.GenomicArray"><code class="xref py py-class docutils literal"><span class="pre">GenomicArray</span></code></a> that deal with the common
special case of overlapping features. This is best explained by an example: Let&#8217;s say, we
have two features, <code class="docutils literal"><span class="pre">&quot;geneA&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;geneB&quot;</span></code>, that are at overlapping positions:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ivA</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ivB</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">400</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>In a <code class="docutils literal"><span class="pre">GenomicArrayOfSets</span></code>, the value of each step is a <code class="docutils literal"><span class="pre">set</span></code> and so can hold more than
one object. The __iadd__ method is overloaded to add elements to the sets:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gas</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArrayOfSets</span><span class="p">(</span> <span class="p">[</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr2&quot;</span><span class="p">],</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gas</span><span class="p">[</span><span class="n">ivA</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot;gene A&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gas</span><span class="p">[</span><span class="n">ivB</span><span class="p">]</span> <span class="o">+=</span> <span class="s2">&quot;gene B&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[(</span><span class="n">st</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">st</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span> <span class="k">for</span> <span class="n">st</span> <span class="ow">in</span> <span class="n">gas</span><span class="p">[</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">()]</span> 
<span class="go">[(&lt;GenomicInterval object &#39;chr1&#39;, [0,100), strand &#39;.&#39;&gt;,   []),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [100,200), strand &#39;.&#39;&gt;, [&#39;gene A&#39;]),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [200,300), strand &#39;.&#39;&gt;, [&#39;gene A&#39;, &#39;gene B&#39;]),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [300,400), strand &#39;.&#39;&gt;, [&#39;gene B&#39;]),</span>
<span class="go"> (&lt;GenomicInterval object &#39;chr1&#39;, [400,500), strand &#39;.&#39;&gt;, [])]</span>
</pre></div>
</div>
<dl class="class">
<dt id="HTSeq.GenomicArrayOfSets">
<em class="property">class </em><code class="descclassname">HTSeq.</code><code class="descname">GenomicArrayOfSets</code><span class="sig-paren">(</span><em>chroms</em>, <em>stranded = True</em><span class="sig-paren">)</span><a class="headerlink" href="#HTSeq.GenomicArrayOfSets" title="Permalink to this definition">¶</a></dt>
<dd><p>Instantiation is as in <a class="reference internal" href="#HTSeq.GenomicArray" title="HTSeq.GenomicArray"><code class="xref py py-class docutils literal"><span class="pre">GenomicArray</span></code></a>, only that <code class="docutils literal"><span class="pre">datatype</span></code> is always <code class="docutils literal"><span class="pre">'O'</span></code>.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Genomic intervals and genomic arrays</a><ul>
<li><a class="reference internal" href="#genomicinterval"><code class="docutils literal"><span class="pre">GenomicInterval</span></code></a></li>
<li><a class="reference internal" href="#genomicposition"><code class="docutils literal"><span class="pre">GenomicPosition</span></code></a></li>
<li><a class="reference internal" href="#genomicarray"><code class="docutils literal"><span class="pre">GenomicArray</span></code></a></li>
<li><a class="reference internal" href="#genomicarrayofsets"><code class="docutils literal"><span class="pre">GenomicArrayOfSets</span></code></a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="sequences.html"
                        title="previous chapter">Sequences and FASTA/FASTQ files</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="alignments.html"
                        title="next chapter">Read alignments</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/genomic.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="alignments.html" title="Read alignments"
             >next</a> |</li>
        <li class="right" >
          <a href="sequences.html" title="Sequences and FASTA/FASTQ files"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>