<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Quality Assessment with htseq-qa &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Counting reads in features with htseq-count" href="count.html" />
    <link rel="prev" title="Miscellaneous" href="misc.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="count.html" title="Counting reads in features with htseq-count"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="misc.html" title="Miscellaneous"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="quality-assessment-with-htseq-qa">
<span id="qa"></span><h1>Quality Assessment with <code class="docutils literal"><span class="pre">htseq-qa</span></code><a class="headerlink" href="#quality-assessment-with-htseq-qa" title="Permalink to this headline">¶</a></h1>
<p>The Python script <code class="docutils literal"><span class="pre">htseq-qa</span></code> takes a file with sequencing reads (either
raw or aligned reads) and produces a PDF file with useful plots to assess
the technical quality of a run.</p>
<div class="section" id="plot">
<h2>Plot<a class="headerlink" href="#plot" title="Permalink to this headline">¶</a></h2>
<p>Here is a typical plot:</p>
<a class="reference internal image-reference" href="_images/qa_example.png"><img alt="_images/qa_example.png" src="_images/qa_example.png" style="width: 600px;" /></a>
<p>The plot is made from a SAM file, which contained aligned and unalignable reads.
The left column is made from the non-aligned, the right column from the aligned
reads. The header informs you about the name of the SAM file, and the number of
reads.</p>
<p>The upper row shows how often which base was called for each position in the
read. In this sample, the non-alignable reads have a clear excess in A. The
aligned reads have a balance between complementing reads: A and C (reddish colours)
have equal levels, and so do C and G (greenish colours). The sequences seem to be AT
rich. Furthermore, nearly all aligned reads start with a T, followed by an A, and then,
a C in 70% and an A in 30% of the reads. Such an imbalance would be reason for concern
if it has no good explanation. Here, the reason is that the fragmentation of the sample
was done by enzyme digestion.</p>
<p>The lower half shows the abundance of base-call quality scores at the different positions
in the read. Nearly all aligned reads have a quality of 34 over their whole length, while
for the non-aligned reads, some reads have lower quality scores towards their ends.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Note that <code class="docutils literal"><span class="pre">htseq-qa</span></code> needs matplotlib to produce the plot, so you need to install this
module, as described <a class="reference external" href="http://matplotlib.sourceforge.net/users/installing.html">here</a> on the matplotlib web site.</p>
<p>After you have installed HTSeq (see <a class="reference internal" href="install.html#install"><span class="std std-ref">Prequisites and installation</span></a>) and matplotlib, you can run <code class="docutils literal"><span class="pre">htseq-qa</span></code> from
the command line:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">htseq</span><span class="o">-</span><span class="n">qa</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="n">read_file</span>
</pre></div>
</div>
<p>If the file <code class="docutils literal"><span class="pre">htseq-qa</span></code> is not in your path, you can, alternatively, call the script with</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="o">-</span><span class="n">m</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">scripts</span><span class="o">.</span><span class="n">qa</span> <span class="p">[</span><span class="n">options</span><span class="p">]</span> <span class="n">read_file</span>
</pre></div>
</div>
<p>The <cite>read_file</cite> is either a FASTQ file or a SAM file. For a SAM file, a plot with two columns
is produced as above, for a FASTQ file, you get only one column.</p>
<p>The output is written into a file with the same name as <cite>read_file</cite>, with the suffix <code class="docutils literal"><span class="pre">.pdf</span></code>
added. View it with a PDF viewer such as the Acrobat Reader.</p>
<div class="section" id="options">
<h3>Options<a class="headerlink" href="#options" title="Permalink to this headline">¶</a></h3>
<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-t">
<code class="descname">-t</code><code class="descclassname"> &lt;type&gt;</code><code class="descclassname">, </code><code class="descname">--type</code><code class="descclassname">=&lt;type&gt;</code><a class="headerlink" href="#cmdoption-htseq-qa-t" title="Permalink to this definition">¶</a></dt>
<dd><p>The file type of the <cite>read_file</cite>. Supported values for <cite>&lt;type&gt;</cite> are:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">sam</span></code>: a SAM file (Note that the <a class="reference external" href="http://samtools.sourceforge.net/">SAMtools</a> contain Perl scripts to convert
most alignment formats to SAM)</li>
<li><code class="docutils literal"><span class="pre">solexa-export</span></code>: an <code class="docutils literal"><span class="pre">_export.txt</span></code> file as produced by the SolexaPipeline
software after aligning with Eland (<code class="docutils literal"><span class="pre">htseq-qa</span></code> expects the new Solexa quality
encoding as produced by version 1.3 or newer of the SolexaPipeline)</li>
<li><code class="docutils literal"><span class="pre">fastq</span></code>: a FASTQ file with standard (Sanger or Phred) quality encoding</li>
<li><code class="docutils literal"><span class="pre">solexa-fastq</span></code>: a FASTQ file with Solexa quality encoding, as produced by
the SolexaPipeline after base-calling with Bustard (<code class="docutils literal"><span class="pre">htseq-qa</span></code> expects
the new Solexa quality encoding as produced by version 1.3 or newer
of the SolexaPipeline)</li>
</ul>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-o">
<code class="descname">-o</code><code class="descclassname"> &lt;outfile&gt;</code><code class="descclassname">, </code><code class="descname">--outfile</code><code class="descclassname">=&lt;outfile&gt;</code><a class="headerlink" href="#cmdoption-htseq-qa-o" title="Permalink to this definition">¶</a></dt>
<dd><p>output filename (default is <cite>&lt;read_file&gt;``</cite>.pdf``)</p>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-r">
<code class="descname">-r</code><code class="descclassname"> &lt;readlen&gt;</code><code class="descclassname">, </code><code class="descname">--readlength</code><code class="descclassname">=&lt;readlen&gt;</code><a class="headerlink" href="#cmdoption-htseq-qa-r" title="Permalink to this definition">¶</a></dt>
<dd><p>the maximum read length (when not specified, the
script guesses from the file</p>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-g">
<code class="descname">-g</code><code class="descclassname"> &lt;gamma&gt;</code><code class="descclassname">, </code><code class="descname">--gamma</code><code class="descclassname">=&lt;gamma&gt;</code><a class="headerlink" href="#cmdoption-htseq-qa-g" title="Permalink to this definition">¶</a></dt>
<dd><p>the gamma factor for the contrast adjustment of the
quality score plot</p>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--nosplit</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-htseq-qa-n" title="Permalink to this definition">¶</a></dt>
<dd><p>do not split reads in unaligned and aligned ones, i.e., produce
a one-column plot</p>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--maxqual</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-htseq-qa-m" title="Permalink to this definition">¶</a></dt>
<dd><p>the maximum quality score that appears in the data (default: 40)</p>
</dd></dl>

<dl class="cmdoption">
<dt id="cmdoption-htseq-qa-h">
<code class="descname">-h</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--help</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-htseq-qa-h" title="Permalink to this definition">¶</a></dt>
<dd><p>Show a usage summary and exit</p>
</dd></dl>

</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quality Assessment with <code class="docutils literal"><span class="pre">htseq-qa</span></code></a><ul>
<li><a class="reference internal" href="#plot">Plot</a></li>
<li><a class="reference internal" href="#usage">Usage</a><ul>
<li><a class="reference internal" href="#options">Options</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="misc.html"
                        title="previous chapter">Miscellaneous</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="count.html"
                        title="next chapter">Counting reads in features with <code class="docutils literal"><span class="pre">htseq-count</span></code></a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/qa.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="count.html" title="Counting reads in features with htseq-count"
             >next</a> |</li>
        <li class="right" >
          <a href="misc.html" title="Miscellaneous"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>