<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>A detailed use case: TSS plots &#8212; HTSeq 0.7.2 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.7.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Counting reads" href="counting.html" />
    <link rel="prev" title="A tour through HTSeq" href="tour.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="counting.html" title="Counting reads"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tour.html" title="A tour through HTSeq"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="a-detailed-use-case-tss-plots">
<span id="tss"></span><h1>A detailed use case: TSS plots<a class="headerlink" href="#a-detailed-use-case-tss-plots" title="Permalink to this headline">¶</a></h1>
<p>A common task in ChIP-Seq analysis is to get profiles of marks with respect to
nearby features. For example, when analysing histone marks, one is often interested
in the position and extend of such marks in the vicinity of transcription start
sites (TSSs). To this end, one commonly calculates the coverage of reads or fragments
across the whole genome, then marks out fixed-size windows centered  around the
TSSs of all genes, takes the coverages in these windows and adds them up to a
&#8220;profile&#8221; that has the size of the window. This is a simple operation, which, however,
can become challenging, when working with large genomes and very many reads.</p>
<p>Here,
we demonstrate various ways of how data flow can be organized in HTSeq by means of
different solutions to this task.</p>
<p>As example data, we use a short excerpt from
the data set by Barski et al. (Cell, 2007, 129:823). We downloaded the FASTQ file
for one of the H3K4me3 samples (Short Read Archive accession number SRR001432),
aligned it against the Homo sapiens genome build GRCh37 with BWA, and provide the
start of this file (actually only containing reads aligned to chromosome 1) as
file <code class="docutils literal"><span class="pre">SRR001432_head.bam</span></code> with the HTSeq example files. As annotation, we use
the file <code class="docutils literal"><span class="pre">Homo_sapiens.GRCh37.56_chrom1.gtf</span></code>, which is the part of the Ensembl
GTF file for Homo sapiens for chromosome 1.</p>
<div class="section" id="using-the-full-coverage">
<h2>Using the full coverage<a class="headerlink" href="#using-the-full-coverage" title="Permalink to this headline">¶</a></h2>
<p>We start with the straight-forward way of calculating the full coverage first
and then summing up the profile. This can be done as described in the <a class="reference internal" href="tour.html#tour"><span class="std std-ref">Tour</span></a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">HTSeq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bamfile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">BAM_Reader</span><span class="p">(</span> <span class="s2">&quot;SRR001432_head.bam&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gtffile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GFF_Reader</span><span class="p">(</span> <span class="s2">&quot;Homo_sapiens.GRCh37.56_chrom1.gtf&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">coverage</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">(</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">typecode</span><span class="o">=</span><span class="s2">&quot;i&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">bamfile</span><span class="p">:</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">almnt</span><span class="o">.</span><span class="n">aligned</span><span class="p">:</span>
<span class="gp">... </span>      <span class="n">coverage</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
<p>To find the location of all transcription start sites, we can look in the GTF
file for exons with exon number 1 (as indicated by the <code class="docutils literal"><span class="pre">exon_number</span></code>
attribute in Ensembl GTF files) and ask for their directional start (<code class="docutils literal"><span class="pre">start_d</span></code>).
The following loop extracts and prints this information (using <code class="docutils literal"><span class="pre">itertools.islice</span></code>
to go through only the first 100 features in the GTF file):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">itertools</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">islice</span><span class="p">(</span> <span class="n">gtffile</span><span class="p">,</span> <span class="mi">100</span><span class="p">):</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
<span class="gp">... </span>      <span class="nb">print</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;gene_id&quot;</span><span class="p">],</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;transcript_id&quot;</span><span class="p">],</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span><span class="p">)</span>
<span class="go">ENSG00000223972 ENST00000456328 1:11873/+</span>
<span class="go">ENSG00000223972 ENST00000450305 1:12009/+</span>
<span class="go">ENSG00000227232 ENST00000423562 1:29369/-</span>
<span class="go">ENSG00000227232 ENST00000438504 1:29369/-</span>
<span class="go">ENSG00000227232 ENST00000488147 1:29569/-</span>
<span class="go">ENSG00000227232 ENST00000430492 1:29342/-</span>
<span class="go">ENSG00000243485 ENST00000473358 1:29553/+</span>
<span class="go">ENSG00000243485 ENST00000469289 1:30266/+</span>
<span class="go">ENSG00000221311 ENST00000408384 1:30365/+</span>
<span class="go">ENSG00000237613 ENST00000417324 1:36080/-</span>
<span class="go">ENSG00000237613 ENST00000461467 1:36072/-</span>
<span class="go">ENSG00000233004 ENST00000421949 1:53048/+</span>
<span class="go">ENSG00000240361 ENST00000492842 1:62947/+</span>
<span class="go">ENSG00000177693 ENST00000326183 1:69054/+</span>
</pre></div>
</div>
<p>As the GTF file contains several transcripts for each gene, one TSS may appear
multiple times, giving undue weight to it. Hence, we collect them in a <code class="docutils literal"><span class="pre">set</span></code>
as this data type enforces uniqueness.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tsspos</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">gtffile</span><span class="p">:</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
<span class="gp">... </span>      <span class="n">tsspos</span><span class="o">.</span><span class="n">add</span><span class="p">(</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span> <span class="p">)</span>
</pre></div>
</div>
<p>Let&#8217;s take one of these starting positions. To get a nice one, we manually chose
this one here, just for demonstration purposes:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">145439814</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span>
</pre></div>
</div>
<p>This is really one of the TSSs in the set:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="ow">in</span> <span class="n">tsspos</span>
<span class="go">True</span>
</pre></div>
</div>
<p>We can get a window centered on this TSS by just subtracting and adding a fixed
value (half of the desired window size, let&#8217;s use 3 kb):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">halfwinwidth</span> <span class="o">=</span> <span class="mi">3000</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">window</span>
<span class="go">&lt;GenomicInterval object &#39;1&#39;, [145436814,145442814), strand &#39;.&#39;&gt;</span>
</pre></div>
</div>
<p>We can check the coverage in this window by subsetting and transforming to a list:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span> <span class="n">coverage</span><span class="p">[</span><span class="n">window</span><span class="p">]</span> <span class="p">)</span>  
<span class="go">[0, 0, 0, ..., 0, 0]</span>
</pre></div>
</div>
<p>As we will work with <a class="reference external" href="http://numpy.scipy.org/">numpy</a> from now on, it may be better to get this as
numpy array:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wincvg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">fromiter</span><span class="p">(</span> <span class="n">coverage</span><span class="p">[</span><span class="n">window</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">wincvg</span>
<span class="go">array([0, 0, 0, ..., 0, 0, 0], dtype=int32)</span>
</pre></div>
</div>
<p>With matplotlib, we can see that this vector is, in effect, not all zero:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span> 
<span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">wincvg</span> <span class="p">)</span>    
<span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>            
</pre></div>
</div>
<img alt="_images/tss_fig1.png" src="_images/tss_fig1.png" />
<p>To sum up the profile, we initialize a numpy vector of the size of our window with zeroes:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">profile</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span> <span class="p">)</span>
</pre></div>
</div>
<p>Now, we can go through the TSS positions and add the coverage in their windows
to get the profile:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">tsspos</span><span class="p">:</span>
<span class="gp">... </span>   <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">... </span>   <span class="n">wincvg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">fromiter</span><span class="p">(</span> <span class="n">coverage</span><span class="p">[</span><span class="n">window</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="p">)</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
<span class="gp">... </span>      <span class="n">profile</span> <span class="o">+=</span> <span class="n">wincvg</span>
<span class="gp">... </span>   <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>      <span class="n">profile</span> <span class="o">+=</span> <span class="n">wincvg</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>Note that we add the window coverage reversed (&#8220;<code class="docutils literal"><span class="pre">[::-1]</span></code>&#8221;) if the gene was on the minus
strand.</p>
<p>Using matplotlib, we can plot this:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profile</span> <span class="p">)</span>  
<span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>  
</pre></div>
</div>
<img alt="_images/tss_fig2.png" src="_images/tss_fig2.png" />
<p>We can see clearly that the reads concentrate around the TSS, with a prominent peak
a bit downstream (if you use matplotlib&#8217;s interactive zoom, you can easily see that
the peak is at 153 bp) and a dip upstream, at -79 bp.</p>
<p>Going back to the beginning, there is one possible improvement: When calculating the
coverage, we just added one to all the base pairs that the read covered. However, the
fragment extends beyond the read, to a length of about 200 bp (the fragment size for
which Barski et al. selected). Maybe we get a better picture by calculating
the coverage not from the reads but from the <em>fragments</em>, i.e., the reads extended
to fragment size. For this, we just one
line, to extend the read to 200 bp. Using this, we now put the whole script together:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">HTSeq</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span>

<span class="n">bamfile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">BAM_Reader</span><span class="p">(</span> <span class="s2">&quot;SRR001432_head.bam&quot;</span> <span class="p">)</span>
<span class="n">gtffile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GFF_Reader</span><span class="p">(</span> <span class="s2">&quot;Homo_sapiens.GRCh37.56_chrom1.gtf&quot;</span> <span class="p">)</span>
<span class="n">halfwinwidth</span> <span class="o">=</span> <span class="mi">3000</span>
<span class="n">fragmentsize</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">coverage</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArray</span><span class="p">(</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">typecode</span><span class="o">=</span><span class="s2">&quot;i&quot;</span> <span class="p">)</span>
<span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">bamfile</span><span class="p">:</span>
   <span class="k">if</span> <span class="n">almnt</span><span class="o">.</span><span class="n">aligned</span><span class="p">:</span>
      <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
      <span class="n">coverage</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="n">tsspos</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">gtffile</span><span class="p">:</span>
   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
      <span class="n">tsspos</span><span class="o">.</span><span class="n">add</span><span class="p">(</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span> <span class="p">)</span>

<span class="n">profile</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span> <span class="p">)</span>      
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">tsspos</span><span class="p">:</span>
   <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
   <span class="n">wincvg</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">fromiter</span><span class="p">(</span> <span class="n">coverage</span><span class="p">[</span><span class="n">window</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="p">)</span>
   <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
      <span class="n">profile</span> <span class="o">+=</span> <span class="n">wincvg</span>
   <span class="k">else</span><span class="p">:</span>
      <span class="n">profile</span> <span class="o">+=</span> <span class="n">wincvg</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

</pre></div>
</div>
<p>The script produces a <code class="docutils literal"><span class="pre">profile</span></code> variable whhich we can plot by adding these lines
to it:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profile</span> <span class="p">)</span>  
<span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>  
</pre></div>
</div>
<img alt="_images/tss_fig3.png" src="_images/tss_fig3.png" />
<p>The plot looks much smoother with the extended fragments.</p>
<p>The coverage vector can be held in memory, even for a very large genome, because large
parts of it are zero and even where there are reads, the values tend to stay constant
for stretches of several bases. Hence, GenomicArray&#8217;s step storage mode is useful.
If, however, extremely large amounts of reads are processed, the coverage vector can become &#8220;rough&#8221; and
change value at every position. Then, the step storage mode becomes inefficient
and we might be better off with an ordinary dense vector such as provided by numpy.
As this numpy vector becomes very large, it may not fit in memory, and the &#8216;memmap&#8217;
storage (using numpy&#8217;s memmap facility) then uses temporary files on disk. We
mention these possibilities as they may be useful when working with the full coverage vector
is required. Here, however, we can do otherwise.</p>
</div>
<div class="section" id="using-indexed-bam-files">
<h2>Using indexed BAM files<a class="headerlink" href="#using-indexed-bam-files" title="Permalink to this headline">¶</a></h2>
<p>We do not need the coverage everythere. We only need it close to the TSSs. We can
sort our BAM file by position (<code class="docutils literal"><span class="pre">samtools</span> <span class="pre">sort</span></code>) and index it (<code class="docutils literal"><span class="pre">samtools</span> <span class="pre">index</span></code>)
and then use random access, as HTSeq exposes this functionality of SAMtools.</p>
<p>Let&#8217;s say we use the same window as above as example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicPosition</span><span class="p">(</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="mi">145439814</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">window</span>
<span class="go">&lt;GenomicInterval object &#39;1&#39;, [145436814,145442814), strand &#39;.&#39;&gt;</span>
</pre></div>
</div>
<p>Then, we can simply get a list of all reads within this interval as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sortedbamfile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">BAM_Reader</span><span class="p">(</span> <span class="s2">&quot;SRR001432_head_sorted.bam&quot;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">sortedbamfile</span><span class="p">[</span> <span class="n">window</span> <span class="p">]:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">almnt</span><span class="p">)</span>   
<span class="go">&lt;SAM_Alignment object: Read &#39;SRR001432.90270 USI-EAS21_0008_3445:8:3:245:279 length=25&#39; aligned to 1:[145437532,145437557)/-&gt;</span>
<span class="go"> ...</span>
<span class="go">&lt;SAM_Alignment object: Read &#39;SRR001432.205754 USI-EAS21_0008_3445:8:5:217:355 length=25&#39; aligned to 1:[145440975,145441000)/-&gt;</span>
</pre></div>
</div>
<p>Let&#8217;s have a closer look at the last alignment. As before, we first extent the read to fragment size:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fragmentsize</span> <span class="o">=</span> <span class="mi">200</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">almnt</span>
<span class="go">&lt;SAM_Alignment object: Read &#39;SRR001432.205754 USI-EAS21_0008_3445:8:5:217:355 length=25&#39; aligned to 1:[145440800,145441000)/-&gt;</span>
</pre></div>
</div>
<p>The read has been aligned to the &#8220;-&#8221;
strand, and hence, we should look at its distance to the <em>end</em> of the window
(i.e., <code class="docutils literal"><span class="pre">p.pos</span></code>, the position of the TSS, plus half the window width)
to see where it should be added to the <code class="docutils literal"><span class="pre">profile</span></code> vector:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">start_in_window</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">start_in_window</span><span class="p">,</span> <span class="n">end_in_window</span><span class="p">)</span>
<span class="go">1814 2014</span>
</pre></div>
</div>
<p>To account for this read, we should add ones in the profile vector along
the indicated interval.</p>
<p>Using this, we can go through the set of all TSS positions (in the <code class="docutils literal"><span class="pre">tsspos</span></code>
set variable that we created above) and for each TSS position, loop through
all aligned reads close to it. Here is this double loop:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">profileB</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">tsspos</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">... </span>    <span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">sortedbamfile</span><span class="p">[</span> <span class="n">window</span> <span class="p">]:</span>
<span class="gp">... </span>        <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span>
<span class="gp">... </span>            <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>   <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span>
<span class="gp">... </span>        <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>
<span class="gp">... </span>            <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span>
<span class="gp">... </span>        <span class="n">profileB</span><span class="p">[</span> <span class="n">start_in_window</span> <span class="p">:</span> <span class="n">end_in_window</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
<p>This loop now runs a good deal faster than our first attempt, and has a much
smaller memory footprint.</p>
<p>We can plot the profiles obtained from our two methods on top of each other:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profile</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span> <span class="p">)</span>   
<span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profileB</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span> <span class="p">)</span>   
<span class="gp">&gt;&gt;&gt; </span><span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   
</pre></div>
</div>
<img alt="_images/tss_fig4.png" src="_images/tss_fig4.png" />
<p>We notice that they are equal, except for the boundaries. This artifact arose
because we extend reads to fragment length: A read which is just outside
the <code class="docutils literal"><span class="pre">window</span></code> will not be found by our new loop even though if may reach into our
profile window after extension to fragment length. Therefore, we should make the
window used to subset the BAM file a bit wider than before to get even reads that
are once the fragment length away. However, with this, we may also get reads that
get extended into the wrong direction, such that <code class="docutils literal"><span class="pre">start_in_windows</span></code> and
<code class="docutils literal"><span class="pre">end_in_windows</span></code> extend beyond the size of the fragment vector. Four extra lines
need to be added to deal with these cases, and then, our new script gives the
same result as the previous one.</p>
<p>Here is the complete code:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">HTSeq</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span>

<span class="n">sortedbamfile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">BAM_Reader</span><span class="p">(</span> <span class="s2">&quot;SRR001432_head_sorted.bam&quot;</span> <span class="p">)</span>
<span class="n">gtffile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GFF_Reader</span><span class="p">(</span> <span class="s2">&quot;Homo_sapiens.GRCh37.56_chrom1.gtf&quot;</span> <span class="p">)</span>
<span class="n">halfwinwidth</span> <span class="o">=</span> <span class="mi">3000</span>
<span class="n">fragmentsize</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">tsspos</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">gtffile</span><span class="p">:</span>
   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
      <span class="n">tsspos</span><span class="o">.</span><span class="n">add</span><span class="p">(</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span> <span class="p">)</span>

<span class="n">profile</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i&#39;</span> <span class="p">)</span>   
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">tsspos</span><span class="p">:</span>
   <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> 
       <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">fragmentsize</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">+</span> <span class="n">fragmentsize</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
   <span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">sortedbamfile</span><span class="p">[</span> <span class="n">window</span> <span class="p">]:</span>
      <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
      <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
         <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> 
         <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>   <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> 
      <span class="k">else</span><span class="p">:</span>
         <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>
         <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span>
      <span class="n">start_in_window</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span> <span class="n">start_in_window</span><span class="p">,</span> <span class="mi">0</span> <span class="p">)</span>
      <span class="n">end_in_window</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span> <span class="n">end_in_window</span><span class="p">,</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="p">)</span>
      <span class="k">if</span> <span class="n">start_in_window</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="ow">or</span> <span class="n">end_in_window</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
         <span class="k">continue</span>
      <span class="n">profile</span><span class="p">[</span> <span class="n">start_in_window</span> <span class="p">:</span> <span class="n">end_in_window</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

</pre></div>
</div>
<p>As before, to get a plot, add:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profile</span> <span class="p">)</span>   
<span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>   
</pre></div>
</div>
<p>You will now get the same plot as we got with the first method.</p>
</div>
<div class="section" id="streaming-through-all-reads">
<h2>Streaming through all reads<a class="headerlink" href="#streaming-through-all-reads" title="Permalink to this headline">¶</a></h2>
<p>The previous solution requires sorting and indexing the BAM file. For large amounts
of data, this may be a burden, and hence, we show a third solution that does not
require random access to reads. The idea is to go through all reads in arbitrary
order, check for each read whether it falls into one or more windows around TSSs,
and, if so, adds ones to the profile vector at the appriate places. In essence, it
is the same tactic as before, but nesting the two <em>for</em> loops the other way round.</p>
<p>In order to be able to check quickly whether a read overlaps with a window, we
can use a <a class="reference internal" href="genomic.html#HTSeq.GenomicArrayOfSets" title="HTSeq.GenomicArrayOfSets"><code class="xref py py-class docutils literal"><span class="pre">GenomicArrayOfSets</span></code></a>, in which we mark off all windows. For
easy access, we denote each winow with an <a class="reference internal" href="genomic.html#HTSeq.GenomicPosition" title="HTSeq.GenomicPosition"><code class="xref py py-class docutils literal"><span class="pre">GenomicPosition</span></code></a> object
giving its midpoint, i.e., the actual TSS position, as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tssarray</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArrayOfSets</span><span class="p">(</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">gtffile</span><span class="p">:</span>
<span class="gp">... </span>   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
<span class="gp">... </span>      <span class="n">p</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span>
<span class="gp">... </span>      <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
<span class="gp">... </span>      <span class="n">tssarray</span><span class="p">[</span> <span class="n">window</span> <span class="p">]</span> <span class="o">+=</span> <span class="n">p</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span> <span class="nb">list</span><span class="p">(</span> <span class="n">tssarray</span><span class="o">.</span><span class="n">chrom_vectors</span><span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">][</span><span class="s2">&quot;.&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">()</span> <span class="p">)</span> <span class="p">)</span>
<span class="go">30085</span>
</pre></div>
</div>
<p>As before, <code class="docutils literal"><span class="pre">p</span></code> is the position of the TSS, and <code class="docutils literal"><span class="pre">window</span></code> is the interval
around it.</p>
<p>To demonstrate how this data structure can be used, we take a specific read that
we selected as a good example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">bamfile</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">almnt</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span> <span class="s2">&quot;SRR001432.700 &quot;</span> <span class="p">):</span>
<span class="gp">... </span>        <span class="k">break</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">almnt</span>
<span class="go">&lt;SAM_Alignment object: Read &#39;SRR001432.700 USI-EAS21_0008_3445:8:1:35:294 length=25&#39; aligned to 1:[169677855,169677880)/-&gt;</span>
</pre></div>
</div>
<p>Again, we extent the read to fragment size:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">almnt</span>
<span class="go">&lt;SAM_Alignment object: Read &#39;SRR001432.700 USI-EAS21_0008_3445:8:1:35:294 length=25&#39; aligned to 1:[169677680,169677880)/-&gt;</span>
</pre></div>
</div>
<p>To see which windows the read covers, we subset the <code class="docutils literal"><span class="pre">tssarray</span></code> and ask for steps
that the fragment in <code class="docutils literal"><span class="pre">almnt</span></code> covers:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">step_iv</span><span class="p">,</span> <span class="n">step_set</span> <span class="ow">in</span> <span class="n">tssarray</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">():</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Step&quot;</span><span class="p">,</span> <span class="n">step_iv</span><span class="p">,</span> <span class="s2">&quot;, contained by these windows:&quot;</span><span class="p">)</span>
<span class="gp">... </span>   <span class="n">out</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="gp">... </span>   <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">step_set</span><span class="p">:</span>
<span class="gp">... </span>       <span class="n">out</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="s2">&quot;   Window around TSS at &quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">))</span>
<span class="gp">... </span>   <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">out</span><span class="p">)))</span>
<span class="go">Step 1:[169677680,169677838)/. , contained by these windows:</span>
<span class="go">   Window around TSS at 1:169677780/-</span>
<span class="go">   Window around TSS at 1:169679672/-</span>
<span class="go">Step 1:[169677838,169677880)/. , contained by these windows:</span>
<span class="go">   Window around TSS at 1:169677780/-</span>
<span class="go">   Window around TSS at 1:169679672/-</span>
<span class="go">   Window around TSS at 1:169680838/-</span>
</pre></div>
</div>
<p>As is typical for GenomicArrayOfSets, some TSSs appear in more than one step. To make
sure that we don&#8217;t count them twice, we take the union of all the step sets (with
the operator <code class="docutils literal"><span class="pre">|=</span></code>, which means in-place union when used for Python sets):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">step_iv</span><span class="p">,</span> <span class="n">step_set</span> <span class="ow">in</span> <span class="n">tssarray</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">():</span>
<span class="gp">... </span>   <span class="n">s</span> <span class="o">|=</span> <span class="p">{</span><span class="n">x</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">step_set</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>  <span class="c1">#</span>
<span class="go">[&quot;&lt;GenomicPosition object &#39;1&#39;:169677780, strand &#39;-&#39;&gt;&quot;,</span>
<span class="go"> &quot;&lt;GenomicPosition object &#39;1&#39;:169679672, strand &#39;-&#39;&gt;&quot;,</span>
<span class="go"> &quot;&lt;GenomicPosition object &#39;1&#39;:169680838, strand &#39;-&#39;&gt;&quot;]</span>
</pre></div>
</div>
<p>For each of the values for <code class="docutils literal"><span class="pre">p</span></code> in <code class="docutils literal"><span class="pre">s</span></code>, we calculate values for <code class="docutils literal"><span class="pre">start_in_window</span></code>
and <code class="docutils literal"><span class="pre">stop_in_window</span></code>, as before, and then add ones in the <code class="docutils literal"><span class="pre">profile</span></code> vector
at the appropriate places.</p>
<p>Putting all this together leads to this script:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">HTSeq</span>
<span class="kn">import</span> <span class="nn">numpy</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span>

<span class="n">bamfile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">BAM_Reader</span><span class="p">(</span> <span class="s2">&quot;SRR001432_head.bam&quot;</span> <span class="p">)</span>
<span class="n">gtffile</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GFF_Reader</span><span class="p">(</span> <span class="s2">&quot;Homo_sapiens.GRCh37.56_chrom1.gtf&quot;</span> <span class="p">)</span>
<span class="n">halfwinwidth</span> <span class="o">=</span> <span class="mi">3000</span>
<span class="n">fragmentsize</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">tsspos</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicArrayOfSets</span><span class="p">(</span> <span class="s2">&quot;auto&quot;</span><span class="p">,</span> <span class="n">stranded</span><span class="o">=</span><span class="kc">False</span> <span class="p">)</span>
<span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">gtffile</span><span class="p">:</span>
   <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span> <span class="ow">and</span> <span class="n">feature</span><span class="o">.</span><span class="n">attr</span><span class="p">[</span><span class="s2">&quot;exon_number&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
      <span class="n">p</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start_d_as_pos</span>
      <span class="n">window</span> <span class="o">=</span> <span class="n">HTSeq</span><span class="o">.</span><span class="n">GenomicInterval</span><span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">-</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span> <span class="p">)</span>
      <span class="n">tsspos</span><span class="p">[</span> <span class="n">window</span> <span class="p">]</span> <span class="o">+=</span> <span class="n">p</span>

<span class="n">profile</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;i&quot;</span> <span class="p">)</span>
<span class="k">for</span> <span class="n">almnt</span> <span class="ow">in</span> <span class="n">bamfile</span><span class="p">:</span>
   <span class="k">if</span> <span class="n">almnt</span><span class="o">.</span><span class="n">aligned</span><span class="p">:</span>
      <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">length</span> <span class="o">=</span> <span class="n">fragmentsize</span>
      <span class="n">s</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
      <span class="k">for</span> <span class="n">step_iv</span><span class="p">,</span> <span class="n">step_set</span> <span class="ow">in</span> <span class="n">tsspos</span><span class="p">[</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span> <span class="p">]</span><span class="o">.</span><span class="n">steps</span><span class="p">():</span>
         <span class="n">s</span> <span class="o">|=</span> <span class="n">step_set</span>
      <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">s</span><span class="p">:</span>
         <span class="k">if</span> <span class="n">p</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
            <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span> <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span>
            <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>   <span class="o">-</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span>
         <span class="k">else</span><span class="p">:</span>
            <span class="n">start_in_window</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">end</span>
            <span class="n">end_in_window</span>   <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">pos</span> <span class="o">+</span> <span class="n">halfwinwidth</span> <span class="o">-</span> <span class="n">almnt</span><span class="o">.</span><span class="n">iv</span><span class="o">.</span><span class="n">start</span>
         <span class="n">start_in_window</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span> <span class="n">start_in_window</span><span class="p">,</span> <span class="mi">0</span> <span class="p">)</span>
         <span class="n">end_in_window</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span> <span class="n">end_in_window</span><span class="p">,</span> <span class="mi">2</span><span class="o">*</span><span class="n">halfwinwidth</span> <span class="p">)</span>
         <span class="n">profile</span><span class="p">[</span> <span class="n">start_in_window</span> <span class="p">:</span> <span class="n">end_in_window</span> <span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>

</pre></div>
</div>
<p>Again, to get a plot (which will look the same as before), add:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pyplot</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span> <span class="o">-</span><span class="n">halfwinwidth</span><span class="p">,</span> <span class="n">halfwinwidth</span> <span class="p">),</span> <span class="n">profile</span> <span class="p">)</span>  
<span class="n">pyplot</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>  
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">A detailed use case: TSS plots</a><ul>
<li><a class="reference internal" href="#using-the-full-coverage">Using the full coverage</a></li>
<li><a class="reference internal" href="#using-indexed-bam-files">Using indexed BAM files</a></li>
<li><a class="reference internal" href="#streaming-through-all-reads">Streaming through all reads</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tour.html"
                        title="previous chapter">A tour through HTSeq</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="counting.html"
                        title="next chapter">Counting reads</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tss.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="counting.html" title="Counting reads"
             >next</a> |</li>
        <li class="right" >
          <a href="tour.html" title="A tour through HTSeq"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">HTSeq 0.7.2 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2010, Simon Anders.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.5.
    </div>
  </body>

<!-- Start of StatCounter Code -->
<script type="text/javascript">
var sc_project=5680199; 
var sc_invisible=1; 
var sc_partition=63; 
var sc_click_stat=1; 
var sc_security="6e579550"; 
</script>

<script type="text/javascript"
src="http://www.statcounter.com/counter/counter.js"></script><noscript><div
class="statcounter"><a title="counter for iweb"
href="http://www.statcounter.com/iweb/" target="_blank"><img
class="statcounter"
src="http://c.statcounter.com/5680199/0/6e579550/1/"
alt="counter for iweb" ></a></div></noscript>
<!-- End of StatCounter Code -->
</html>